<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Antigoni Kaliontzopoulou &amp; Michael Collyer. Adapted by Thomas Smith">

<title>An introduction to Geomorph</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="An_introduction_to_Geomorph_files/libs/clipboard/clipboard.min.js"></script>
<script src="An_introduction_to_Geomorph_files/libs/quarto-html/quarto.js"></script>
<script src="An_introduction_to_Geomorph_files/libs/quarto-html/popper.min.js"></script>
<script src="An_introduction_to_Geomorph_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="An_introduction_to_Geomorph_files/libs/quarto-html/anchor.min.js"></script>
<link href="An_introduction_to_Geomorph_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="An_introduction_to_Geomorph_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="An_introduction_to_Geomorph_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="An_introduction_to_Geomorph_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="An_introduction_to_Geomorph_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">An introduction to Geomorph</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Antigoni Kaliontzopoulou &amp; Michael Collyer. Adapted by Thomas Smith </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="preparing-your-data" class="level2">
<h2 class="anchored" data-anchor-id="preparing-your-data">1.0 Preparing your data</h2>
<p>In geomorph, you can digitize fixed landmarks, and semilandmarks on surfaces, using a 3D surface object or a mesh3d object. In its simplest version, this involves the following geomorph functions: digit.fixed, buildtemplate, and digitsurface. If you want to use curve semilandmarks, these may be either digitized as fixed landmarks, and then indexed for sliding using the curves argument of gpagen(); or sampled automatically from a digitized curve using digit.curves().</p>
<section id="digitizing-fixed-landmarksy" class="level3">
<h3 class="anchored" data-anchor-id="digitizing-fixed-landmarksy">1.1 Digitizing fixed landmarksy</h3>
<p>We will be using the scallop example ply file included in geomorph.</p>
<p>First, check you have geomorph installed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">c</span>(<span class="st">"geomorph"</span>)[<span class="sc">!</span><span class="fu">c</span>(<span class="st">"geomorph"</span>) <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[,<span class="st">"Package"</span>]]) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="fu">c</span>(<span class="st">"geomorph"</span>)[<span class="sc">!</span><span class="fu">c</span>(<span class="st">"geomorph"</span>) <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[,<span class="st">"Package"</span>]])</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, let’s load it and the example data in.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geomorph)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: RRPP</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: rgl</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Matrix</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"scallopPLY"</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>my.ply <span class="ot">&lt;-</span> scallopPLY<span class="sc">$</span>ply</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let´s also make sure that our working directory is set to an adequate location. This is particularly relevant for 3D digitizing, as throughout the process, files with the digitized landmark coordinates, as well as the template we will create for replicating 3D surface semilandmarks across specimens, will be stored and automatically read in by functions. For this example, we will work in the subdirectory ~/example3D/, so go ahead and make sure your working directory is set to a folder with that name. I’d recommend creating a new folder on your desktop.</p>
<p>Once a ply file is imported (usually using read.ply), one will first digitize the fixed landmarks using digit.fixed(). Note that this is an interactive function, so you will need to replicate these steps in your R console to be able to see the full version of this example. We will digitize 5 fixed landmarks (you set a landmark by right-clicking, then clicking ‘y’).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># digitize landmarks</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>fixed.lms1 <span class="ot">&lt;-</span> <span class="fu">digit.fixed</span>(<span class="at">spec =</span> my.ply, <span class="at">fixed =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Keep Landmark  1 (y/n)? 
Keep Landmark  2 (y/n)? 
Keep Landmark  3 (y/n)? 
Keep Landmark  4 (y/n)? 
Keep Landmark  5 (y/n)? </code></pre>
</div>
</div>
<p>Note that landmark selection on the 3dmesh is done by choosing the point of the scan closest to where you click with your mouse. This means that occasionally points may appear to “go through to the other side of the structure”, especially if your initial point cloud is not very dense. To solve this issue, try rotating the structure to find a better view for digitizing, or zooming in such that you can click close to one of the scan points (these are visible as tiny grey points when you zoom in). Place your landmarks in the points specified in the support document.</p>
<p>Now have a look at your working directory. A new file has been created, called my.ply.nts. This contains the coordinates of the digitized 3D fixed landmarks for your specimen.</p>
<p>Now let´s add a second specimen. For the example, we will just replicate the ply file of the first one, but with a different name. Go ahead and digitize the same five landmarks on this second specimen. A second nts file, called my.ply.2.nts is now created in your working directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>my.ply<span class="fl">.2</span> <span class="ot">&lt;-</span> scallopPLY<span class="sc">$</span>ply</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>fixed.lms2 <span class="ot">&lt;-</span> <span class="fu">digit.fixed</span>(my.ply<span class="fl">.2</span>, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Keep Landmark  1 (y/n)? 
Keep Landmark  2 (y/n)? 
Keep Landmark  3 (y/n)? 
Keep Landmark  4 (y/n)? 
Keep Landmark  5 (y/n)? </code></pre>
</div>
</div>
<p>If one wishes to continue and sample also surface semilandmarks, this step can be performed together with the acquisition of surface sliders (see below).</p>
</section>
<section id="sampling-surface-semilandmarks" class="level3">
<h3 class="anchored" data-anchor-id="sampling-surface-semilandmarks">1.2 Sampling surface semilandmarks</h3>
<section id="building-the-template" class="level4">
<h4 class="anchored" data-anchor-id="building-the-template">1.2.1 Building the template</h4>
<p>Now that the fixed landmarks have been digitized, we can continue and sample surface sliding semilandmarks. We will use the first specimen as the reference from which to build a template, which will then be used to sample these semilandmarks across all specimens in our sample. This is done using the function buildtemplate. We will sample 100 surface semilandmarks.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>surf.pts1 <span class="ot">&lt;-</span> <span class="fu">buildtemplate</span>(<span class="at">spec =</span> my.ply, <span class="at">fixed =</span> fixed.lms1, <span class="at">surface.sliders =</span> <span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>One may want to import fixed landmarks already digitized and saved as an nts file. This can be easily done using readland.nts, but remember that the input to the argument fixed above is a matrix, so you will need to combine that to two.d.array to obtain a 2d-matrix with the coordinates of the fixed landmarks.</p>
<p>When running buildtemplate you will see the sampled surface semilandmarks in an rgl device. Have a look at it - is it covered in blue dots?</p>
<p>Note that a txt file containing the coordinates of the sampled surface semilandmarks has been created in your working directory, with the name template.txt. Also, note that the file my.ply.nts has been modified, and it now contains the coordinates of the fixed landmarks AND surface semilandmarks. Finally, a csv file containing the index of sliding semilandmarks has been created, and can be read for use with the argument surfaces of gpagen.</p>
<p>One may choose to perform steps 1. and 2. above at once for the reference specimen. In that case, the argument fixed of buildtemplate would indicate the number of fixed landmarks to be digitized (i.e.&nbsp;5 in our example).</p>
</section>
<section id="sampling-surface-semilandmarks-in-subsequent-specimens" class="level4">
<h4 class="anchored" data-anchor-id="sampling-surface-semilandmarks-in-subsequent-specimens">1.2.2 Sampling surface semilandmarks in subsequent specimens</h4>
<p>Now that the template has been created, we will use it to sample homologous surface semilandmarks in the second specimen in our sample. This is done using the function digitsurface.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>surf.pts2 <span class="ot">&lt;-</span> <span class="fu">digitsurface</span>(<span class="at">spec =</span> my.ply<span class="fl">.2</span>, <span class="at">fixed =</span> fixed.lms2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Warping template

  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |                                                                      |   1%
  |                                                                            
  |=                                                                     |   1%
  |                                                                            
  |=                                                                     |   2%
  |                                                                            
  |==                                                                    |   2%
  |                                                                            
  |==                                                                    |   3%
  |                                                                            
  |===                                                                   |   4%
  |                                                                            
  |===                                                                   |   5%
  |                                                                            
  |====                                                                  |   5%
  |                                                                            
  |====                                                                  |   6%
  |                                                                            
  |=====                                                                 |   7%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |======                                                                |   8%
  |                                                                            
  |======                                                                |   9%
  |                                                                            
  |=======                                                               |   9%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=======                                                               |  11%
  |                                                                            
  |========                                                              |  11%
  |                                                                            
  |========                                                              |  12%
  |                                                                            
  |=========                                                             |  12%
  |                                                                            
  |=========                                                             |  13%
  |                                                                            
  |==========                                                            |  14%
  |                                                                            
  |==========                                                            |  15%
  |                                                                            
  |===========                                                           |  15%
  |                                                                            
  |===========                                                           |  16%
  |                                                                            
  |============                                                          |  17%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |=============                                                         |  18%
  |                                                                            
  |=============                                                         |  19%
  |                                                                            
  |==============                                                        |  19%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |==============                                                        |  21%
  |                                                                            
  |===============                                                       |  21%
  |                                                                            
  |===============                                                       |  22%
  |                                                                            
  |================                                                      |  22%
  |                                                                            
  |================                                                      |  23%
  |                                                                            
  |=================                                                     |  24%
  |                                                                            
  |=================                                                     |  25%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |==================                                                    |  26%
  |                                                                            
  |===================                                                   |  27%
  |                                                                            
  |===================                                                   |  28%
  |                                                                            
  |====================                                                  |  28%
  |                                                                            
  |====================                                                  |  29%
  |                                                                            
  |=====================                                                 |  29%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=====================                                                 |  31%
  |                                                                            
  |======================                                                |  31%
  |                                                                            
  |======================                                                |  32%
  |                                                                            
  |=======================                                               |  32%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |========================                                              |  34%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |=========================                                             |  35%
  |                                                                            
  |=========================                                             |  36%
  |                                                                            
  |==========================                                            |  37%
  |                                                                            
  |==========================                                            |  38%
  |                                                                            
  |===========================                                           |  38%
  |                                                                            
  |===========================                                           |  39%
  |                                                                            
  |============================                                          |  39%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |============================                                          |  41%
  |                                                                            
  |=============================                                         |  41%
  |                                                                            
  |=============================                                         |  42%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |==============================                                        |  43%
  |                                                                            
  |===============================                                       |  44%
  |                                                                            
  |===============================                                       |  45%
  |                                                                            
  |================================                                      |  45%
  |                                                                            
  |================================                                      |  46%
  |                                                                            
  |=================================                                     |  47%
  |                                                                            
  |=================================                                     |  48%
  |                                                                            
  |==================================                                    |  48%
  |                                                                            
  |==================================                                    |  49%
  |                                                                            
  |===================================                                   |  49%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |===================================                                   |  51%
  |                                                                            
  |====================================                                  |  51%
  |                                                                            
  |====================================                                  |  52%
  |                                                                            
  |=====================================                                 |  52%
  |                                                                            
  |=====================================                                 |  53%
  |                                                                            
  |======================================                                |  54%
  |                                                                            
  |======================================                                |  55%
  |                                                                            
  |=======================================                               |  55%
  |                                                                            
  |=======================================                               |  56%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |========================================                              |  58%
  |                                                                            
  |=========================================                             |  58%
  |                                                                            
  |=========================================                             |  59%
  |                                                                            
  |==========================================                            |  59%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |==========================================                            |  61%
  |                                                                            
  |===========================================                           |  61%
  |                                                                            
  |===========================================                           |  62%
  |                                                                            
  |============================================                          |  62%
  |                                                                            
  |============================================                          |  63%
  |                                                                            
  |=============================================                         |  64%
  |                                                                            
  |=============================================                         |  65%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |==============================================                        |  66%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |===============================================                       |  68%
  |                                                                            
  |================================================                      |  68%
  |                                                                            
  |================================================                      |  69%
  |                                                                            
  |=================================================                     |  69%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |=================================================                     |  71%
  |                                                                            
  |==================================================                    |  71%
  |                                                                            
  |==================================================                    |  72%
  |                                                                            
  |===================================================                   |  72%
  |                                                                            
  |===================================================                   |  73%
  |                                                                            
  |====================================================                  |  74%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |=====================================================                 |  75%
  |                                                                            
  |=====================================================                 |  76%
  |                                                                            
  |======================================================                |  77%
  |                                                                            
  |======================================================                |  78%
  |                                                                            
  |=======================================================               |  78%
  |                                                                            
  |=======================================================               |  79%
  |                                                                            
  |========================================================              |  79%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |========================================================              |  81%
  |                                                                            
  |=========================================================             |  81%
  |                                                                            
  |=========================================================             |  82%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |==========================================================            |  83%
  |                                                                            
  |===========================================================           |  84%
  |                                                                            
  |===========================================================           |  85%
  |                                                                            
  |============================================================          |  85%
  |                                                                            
  |============================================================          |  86%
  |                                                                            
  |=============================================================         |  87%
  |                                                                            
  |=============================================================         |  88%
  |                                                                            
  |==============================================================        |  88%
  |                                                                            
  |==============================================================        |  89%
  |                                                                            
  |===============================================================       |  89%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |===============================================================       |  91%
  |                                                                            
  |================================================================      |  91%
  |                                                                            
  |================================================================      |  92%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |=================================================================     |  93%
  |                                                                            
  |==================================================================    |  94%
  |                                                                            
  |==================================================================    |  95%
  |                                                                            
  |===================================================================   |  95%
  |                                                                            
  |===================================================================   |  96%
  |                                                                            
  |====================================================================  |  97%
  |                                                                            
  |====================================================================  |  98%
  |                                                                            
  |===================================================================== |  98%
  |                                                                            
  |===================================================================== |  99%
  |                                                                            
  |======================================================================|  99%
  |                                                                            
  |======================================================================| 100%</code></pre>
</div>
</div>
<p>Once the surface semilandmarks have been sampled, you will see a graph with their position in specimen 2. As before, the file my.ply.2.nts has been modified, and it now contains the coordinates of the fixed landmarks (red) AND surface semilandmarks for specimen 2 (green).</p>
<p>Same as for building the template, one may choose to simultaneously digitize fixed landmarks and sample surface semilandmarks using the template.</p>
</section>
</section>
<section id="plotting-3d-specimens" class="level3">
<h3 class="anchored" data-anchor-id="plotting-3d-specimens">1.3 Plotting 3D specimens</h3>
<p>At any moment, you can plot a specimen together with fixed landmarks and surface sliding semilandmarks using plotspec.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotspec</span>(<span class="at">spec =</span> my.ply, <span class="at">digitspec =</span> surf.pts1, <span class="at">fixed =</span> <span class="dv">5</span>, <span class="at">centered =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="principal-components-analysis-and-visualisation-tools-for-exploring-shape-space" class="level2">
<h2 class="anchored" data-anchor-id="principal-components-analysis-and-visualisation-tools-for-exploring-shape-space">2.0 Principal components analysis and visualisation tools for exploring shape space</h2>
<p>Version 3.1.0 of geomorph introduced the function gm.prcomp, and related utility functions (summary() and plot()), for performing principal components analyses on Procrustes shape variables for a set of aligned specimens. This function now includes several different types of analytical options and, combined with other visualization tools available in geomorph, provides tools for exploring variation in shape space.</p>
<p>Throughout, we will be using shape data of several Plethodon species as an example, so let´s first load and superimpose those.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list=</span><span class="fu">ls</span>())</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"plethspecies"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>Y.gpa <span class="ot">&lt;-</span> <span class="fu">gpagen</span>(plethspecies<span class="sc">$</span>land, <span class="at">print.progress =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="traditional-pca-and-visualising-shape-patterns" class="level3">
<h3 class="anchored" data-anchor-id="traditional-pca-and-visualising-shape-patterns">2.1 Traditional PCA and visualising shape patterns</h3>
<p>One first option is to perform a “traditional” PCA, i.e.&nbsp;based on OLS-centering and projection of the data. This is very much like what is performed in the basic R function prcomp.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>PCA <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PCA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Principal Component Analysis 
Centering by OLS mean
Orthogonal projection of OLS residuals
Number of observations: 9 
Number of vectors 8 

Importance of Components:
                              Comp1        Comp2        Comp3        Comp4
Eigenvalues            0.0002720474 0.0001120524 0.0001084758 0.0000568924
Proportion of Variance 0.4564029477 0.1879858086 0.1819855633 0.0954461044
Cumulative Proportion  0.4564029477 0.6443887563 0.8263743196 0.9218204240
                              Comp5        Comp6        Comp7        Comp8
Eigenvalues            0.0000264508 1.260516e-05 5.959622e-06 1.584785e-06
Proportion of Variance 0.0443754550 2.114717e-02 9.998220e-03 2.658730e-03
Cumulative Proportion  0.9661958790 9.873431e-01 9.973413e-01 1.000000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCA, <span class="at">main =</span> <span class="st">"PCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Let’s customise this plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCA, <span class="at">main =</span> <span class="st">"PCA"</span>, <span class="at">pch =</span> <span class="dv">22</span>, <span class="at">bg =</span> <span class="st">"green"</span>, <span class="at">cex =</span> <span class="fl">1.5</span>, <span class="at">cex.lab =</span> <span class="fl">1.5</span>, <span class="at">font.lab =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>One then has several solutions for exploring shape variation across PC space and visualizing shape patterns. First, the user may choose to manually produce deformation grids to compare the shapes corresponding to the extremes of a chosen PC axis using plotRefToTarget. For example, one can:</p>
<ol type="1">
<li>Compare the minimum and maximum values to the global consensus:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>msh <span class="ot">&lt;-</span> <span class="fu">mshape</span>(Y.gpa<span class="sc">$</span>coords)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Minimum</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRefToTarget</span>(PCA<span class="sc">$</span>shapes<span class="sc">$</span>shapes.comp1<span class="sc">$</span>min, msh)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRefToTarget</span>(msh, PCA<span class="sc">$</span>shapes<span class="sc">$</span>shapes.comp1<span class="sc">$</span>max)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Maximum</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Comparing the minimum and maximum values:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRefToTarget</span>(PCA<span class="sc">$</span>shapes<span class="sc">$</span>shapes.comp1<span class="sc">$</span>min, PCA<span class="sc">$</span>shapes<span class="sc">$</span>shapes.comp1<span class="sc">$</span>max, <span class="at">method =</span> <span class="st">"vector"</span>, <span class="at">mag =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Of course here one can use all the plotting options available in plotRefToTarget. Please see the help file of that function for details.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>?plotRefToTarget</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="phylomorphospace" class="level3">
<h3 class="anchored" data-anchor-id="phylomorphospace">2.2 Phylomorphospace</h3>
<p>One may also want to project a phylogeny (if dealing with species-level observations), and estimated ancestral states into the ordination plot produced before, to obtain what is commonly referred to as a “phylomorphospace” plot. This can be easily done by providing a phylogenetic tree during the analysis (which is however NOT used in the analytical step in this case, other than for estimating ancestral states), and then indicating that the phylogeny should also be plotted during the plotting step. For the Plethodon example data, we may project the phylogeny into the previous ordination plot as such:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>PCA.w.phylo <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PCA.w.phylo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Principal Component Analysis 
Centering by OLS mean
Orthogonal projection of OLS residuals
Number of observations: 9 
Number of vectors 8 

Importance of Components:
                              Comp1        Comp2        Comp3        Comp4
Eigenvalues            0.0002720474 0.0001120524 0.0001084758 0.0000568924
Proportion of Variance 0.4564029477 0.1879858086 0.1819855633 0.0954461044
Cumulative Proportion  0.4564029477 0.6443887563 0.8263743196 0.9218204240
                              Comp5        Comp6        Comp7        Comp8
Eigenvalues            0.0000264508 1.260516e-05 5.959622e-06 1.584785e-06
Proportion of Variance 0.0443754550 2.114717e-02 9.998220e-03 2.658730e-03
Cumulative Proportion  0.9661958790 9.873431e-01 9.973413e-01 1.000000e+00


Dispersion (variance) of points, after projection:
                                       Comp1        Comp2        Comp3
Tips Dispersion                 2.720474e-04 1.120524e-04 1.084758e-04
Proportion Tips Dispersion      4.564029e-01 1.879858e-01 1.819856e-01
Cumulative Tips Dispersion      4.564029e-01 6.443888e-01 8.263743e-01
Ancestors Dispersion            3.073342e-05 2.320038e-05 2.039005e-05
Proportion Ancestors Dispersion 3.854327e-01 2.909597e-01 2.557149e-01
Cumulative Ancestors Dispersion 3.854327e-01 6.763923e-01 9.321072e-01
                                       Comp4        Comp5        Comp6
Tips Dispersion                 5.689240e-05 2.645080e-05 1.260516e-05
Proportion Tips Dispersion      9.544610e-02 4.437545e-02 2.114717e-02
Cumulative Tips Dispersion      9.218204e-01 9.661959e-01 9.873431e-01
Ancestors Dispersion            3.171758e-06 1.748508e-06 3.945777e-07
Proportion Ancestors Dispersion 3.977752e-02 2.192831e-02 4.948462e-03
Cumulative Ancestors Dispersion 9.718847e-01 9.938130e-01 9.987615e-01
                                       Comp7        Comp8
Tips Dispersion                 5.959622e-06 1.584785e-06
Proportion Tips Dispersion      9.998220e-03 2.658730e-03
Cumulative Tips Dispersion      9.973413e-01 1.000000e+00
Ancestors Dispersion            8.354868e-08 1.520561e-08
Proportion Ancestors Dispersion 1.047797e-03 1.906960e-04
Cumulative Ancestors Dispersion 9.998093e-01 1.000000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCA.w.phylo, <span class="at">phylo =</span> <span class="cn">TRUE</span>, <span class="at">main =</span> <span class="st">"PCA.w.phylo"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Note that the summary statistics obtained for this analysis are identical to those from the previous one. This is because here the phylogeny is merely used for plotting, and is NOT considered during the analytical procedures.</p>
<p>Again, all plotting arguments can be directly manipulated by the user. Please see the help file of plot.gm.prcomp for details.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>?plot.gm.prcomp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="phylopca" class="level3">
<h3 class="anchored" data-anchor-id="phylopca">2.3 phyloPCA</h3>
<p>Here, the phylogeny IS considered during the analytical step of the ordination, as the principal components analysis is in this case calculated based on GLS-centering and projection of the data. This is to account for the nonindependence among observations of species. For details on the analytical part of this method, see Revell 2009, Evolution 63: 3258 - 3268; Polly et al 2013, Hystrix 24: 33 - 41; Collyer &amp; Adams, submitted.</p>
<p>For the Plethodon example data, this analysis would be implemented and plotted as follows (first with untransformed residual projection, second with transformed residual projection):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Phylo PCA without projecting untransformed residuals</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>phylo.PCA <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy, <span class="at">GLS =</span> <span class="cn">TRUE</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(phylo.PCA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Principal Component Analysis 
Centering by GLS mean
Oblique projection of GLS-centered residuals
Number of observations: 9 
Number of vectors 8 

Importance of Components:
                              Comp1        Comp2        Comp3        Comp4
Eigenvalues            2.761405e-05 9.322868e-06 7.889847e-06 6.605120e-06
Proportion of Variance 4.855705e-01 1.639351e-01 1.387365e-01 1.161457e-01
Cumulative Proportion  4.855705e-01 6.495056e-01 7.882421e-01 9.043878e-01
                              Comp5        Comp6        Comp7        Comp8
Eigenvalues            3.017886e-06 1.203779e-06 9.895110e-07 2.262231e-07
Proportion of Variance 5.306707e-02 2.116747e-02 1.739975e-02 3.977950e-03
Cumulative Proportion  9.574548e-01 9.786223e-01 9.960221e-01 1.000000e+00


Dispersion (variance) of points, after projection:
                                       Comp1        Comp2        Comp3
Tips Dispersion                 0.0002498092 8.395185e-05 1.057114e-04
Proportion Tips Dispersion      0.4190949497 1.408427e-01 1.773479e-01
Cumulative Tips Dispersion      0.4190949497 5.599376e-01 7.372855e-01
Ancestors Dispersion            0.0000169982 9.862682e-06 2.194379e-05
Proportion Ancestors Dispersion 0.2131770701 1.236895e-01 2.752006e-01
Cumulative Ancestors Dispersion 0.2131770701 3.368665e-01 6.120671e-01
                                       Comp4        Comp5        Comp6
Tips Dispersion                 0.0001027383 2.765308e-05 1.675548e-05
Proportion Tips Dispersion      0.1723598707 4.639247e-02 2.811000e-02
Cumulative Tips Dispersion      0.9096453329 9.560378e-01 9.841478e-01
Ancestors Dispersion            0.0000272722 1.114653e-06 1.772175e-06
Proportion Ancestors Dispersion 0.3420249711 1.397905e-02 2.222513e-02
Cumulative Ancestors Dispersion 0.9540920945 9.680711e-01 9.902963e-01
                                       Comp7        Comp8
Tips Dispersion                 7.598972e-06 1.850019e-06
Proportion Tips Dispersion      1.274849e-02 3.103703e-03
Cumulative Tips Dispersion      9.968963e-01 1.000000e+00
Ancestors Dispersion            7.068511e-07 6.689930e-08
Proportion Ancestors Dispersion 8.864732e-03 8.389947e-04
Cumulative Ancestors Dispersion 9.991610e-01 1.000000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(phylo.PCA, <span class="at">phylo =</span> <span class="cn">TRUE</span>, <span class="at">main =</span> <span class="st">"phylo PCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the transformed residuals</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>phylo.tPCA <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy, <span class="at">GLS =</span> <span class="cn">TRUE</span>, <span class="at">transform =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(phylo.tPCA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Principal Component Analysis 
Centering by GLS mean
GLS residuals transformed for orthogonal projection
Number of observations: 9 
Number of vectors 8 

Importance of Components:
                              Comp1        Comp2        Comp3        Comp4
Eigenvalues            2.761405e-05 9.322868e-06 7.889847e-06 6.605120e-06
Proportion of Variance 4.855705e-01 1.639351e-01 1.387365e-01 1.161457e-01
Cumulative Proportion  4.855705e-01 6.495056e-01 7.882421e-01 9.043878e-01
                              Comp5        Comp6        Comp7        Comp8
Eigenvalues            3.017886e-06 1.203779e-06 9.895110e-07 2.262231e-07
Proportion of Variance 5.306707e-02 2.116747e-02 1.739975e-02 3.977950e-03
Cumulative Proportion  9.574548e-01 9.786223e-01 9.960221e-01 1.000000e+00


Dispersion (variance) of points, after projection:
                                       Comp1        Comp2        Comp3
Tips Dispersion                 2.740155e-05 8.385080e-06 7.889441e-06
Proportion Tips Dispersion      4.933393e-01 1.509655e-01 1.420420e-01
Cumulative Tips Dispersion      4.933393e-01 6.443048e-01 7.863468e-01
Ancestors Dispersion            1.604482e-06 8.942699e-07 1.578102e-06
Proportion Ancestors Dispersion 3.034061e-01 1.691056e-01 2.984176e-01
Cumulative Ancestors Dispersion 3.034061e-01 4.725117e-01 7.709293e-01
                                       Comp4        Comp5        Comp6
Tips Dispersion                 6.590826e-06 2.909294e-06 1.175603e-06
Proportion Tips Dispersion      1.186617e-01 5.237912e-02 2.116563e-02
Cumulative Tips Dispersion      9.050085e-01 9.573876e-01 9.785532e-01
Ancestors Dispersion            1.007063e-06 6.196138e-08 1.114791e-07
Proportion Ancestors Dispersion 1.904346e-01 1.171684e-02 2.108060e-02
Cumulative Ancestors Dispersion 9.613640e-01 9.730808e-01 9.941614e-01
                                       Comp7        Comp8
Tips Dispersion                 9.878523e-07 2.033663e-07
Proportion Tips Dispersion      1.778536e-02 3.661419e-03
Cumulative Tips Dispersion      9.963386e-01 1.000000e+00
Ancestors Dispersion            2.711996e-08 3.755919e-09
Proportion Ancestors Dispersion 5.128360e-03 7.102409e-04
Cumulative Ancestors Dispersion 9.992898e-01 1.000000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(phylo.tPCA, <span class="at">phylo =</span> <span class="cn">TRUE</span>, <span class="at">main =</span> <span class="st">"phylo PCA with transformed projection"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>What looks different?</p>
</section>
<section id="paca-phylogenetically-aligned-pca" class="level3">
<h3 class="anchored" data-anchor-id="paca-phylogenetically-aligned-pca">2.4 PaCA: phylogenetically-aligned PCA</h3>
<p>This recently introduced method (Collyer &amp; Adams, submitted) provides an ordination that aligns phenotypic data with phylogenetic signal, by maximizing variation in directions that describe phylogenetic signal, while simultaneously preserving the Euclidean distances among observations in the data space. PaCA provides a projection that shows the most phylogenetic signal in the first few components, irrespective of other signals in the data. By comparing PCA, phyloPCA and PaCA results, one may glean the relative importance of phylogenetic and other (ecological) signals in the data.</p>
<p>For the Plethodon example data, this analysis would be implemented and plotted as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>PaCA <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy, <span class="at">align.to.phy =</span> <span class="cn">TRUE</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PaCA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Alignment to an alternative matrix 
Alignment matrix: phy 
Centering by OLS mean
OLS residuals
Alignment to  phy means residual projection is not orthogonal.
Number of observations: 9 
Number of vectors 8 

Importance of Components:
                                Comp1        Comp2        Comp3        Comp4
Singular Value           0.0003931629 0.0001155198 5.091258e-05 2.660786e-05
Proportion of Covariance 0.6595936040 0.1938028696 8.541399e-02 4.463894e-02
Cumulative Proportion    0.6595936040 0.8533964737 9.388105e-01 9.834494e-01
RV by Component          0.0993211764 0.0291827102 1.286158e-02 6.721702e-03
Cumulative RV            0.0993211764 0.1285038867 1.413655e-01 1.480872e-01
                                Comp5        Comp6        Comp7        Comp8
Singular Value           6.598675e-06 2.351469e-06 7.108871e-07 2.042572e-07
Proportion of Covariance 1.107033e-02 3.944966e-03 1.192627e-03 3.426742e-04
Cumulative Proportion    9.945197e-01 9.984647e-01 9.996573e-01 1.000000e+00
RV by Component          1.666963e-03 5.940304e-04 1.795850e-04 5.159965e-05
Cumulative RV            1.497541e-01 1.503482e-01 1.505278e-01 1.505794e-01


Dispersion (variance) of points, after projection:
                                       Comp1        Comp2        Comp3
Tips Dispersion                 2.301865e-04 1.124810e-04 1.133102e-04
Proportion Tips Dispersion      3.861747e-01 1.887049e-01 1.900960e-01
Cumulative Tips Dispersion      3.861747e-01 5.748795e-01 7.649755e-01
Ancestors Dispersion            4.069158e-05 2.437196e-05 8.067595e-06
Proportion Ancestors Dispersion 5.103196e-01 3.056526e-01 1.011770e-01
Cumulative Ancestors Dispersion 5.103196e-01 8.159722e-01 9.171492e-01
                                       Comp4        Comp5        Comp6
Tips Dispersion                 7.526878e-05 4.046474e-05 1.571235e-05
Proportion Tips Dispersion      1.262754e-01 6.788606e-02 2.635999e-02
Cumulative Tips Dispersion      8.912509e-01 9.591370e-01 9.854970e-01
Ancestors Dispersion            6.244464e-06 3.539330e-07 3.508799e-09
Proportion Ancestors Dispersion 7.831282e-02 4.438730e-03 4.400441e-05
Cumulative Ancestors Dispersion 9.954620e-01 9.999008e-01 9.999448e-01
                                       Comp7        Comp8
Tips Dispersion                 6.396433e-06 2.248360e-06
Proportion Tips Dispersion      1.073104e-02 3.771984e-03
Cumulative Tips Dispersion      9.962280e-01 1.000000e+00
Ancestors Dispersion            2.594059e-09 1.810230e-09
Proportion Ancestors Dispersion 3.253251e-05 2.270238e-05
Cumulative Ancestors Dispersion 9.999773e-01 1.000000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PaCA, <span class="at">phylo =</span> <span class="cn">TRUE</span>, <span class="at">main =</span> <span class="st">"PaCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="three-dimensional-pca-plot-with-a-phylogeny-and-time-on-the-z-axis" class="level3">
<h3 class="anchored" data-anchor-id="three-dimensional-pca-plot-with-a-phylogeny-and-time-on-the-z-axis">2.5 Three-dimensional PCA plot with a phylogeny and time on the z-axis</h3>
<p>Finally, plot.gm.prcomp provides the possibility of producing a 3D plot of any two PCA axes, with the phylogenetic tree connecting the observations and time on the z-axis. Again, different plotting parameters can be controlled to manipulate plot aesthetics. Note, that in this case an rgl plotting device will open for the 3D plot, but the corresponding biplot with the phylogeny projected (option 2, above) will also be produced.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCA.w.phylo, <span class="at">time.plot =</span> <span class="cn">TRUE</span>, <span class="at">pch =</span> <span class="dv">22</span>, <span class="at">bg =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"red"</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="st">"green"</span>, <span class="dv">4</span>)), <span class="at">cex =</span> <span class="dv">2</span>, </span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">phylo.par =</span> <span class="fu">list</span>(<span class="at">edge.color =</span> <span class="st">"grey60"</span>, <span class="at">edge.width =</span> <span class="fl">1.5</span>, <span class="at">tip.txt.cex =</span> <span class="fl">0.75</span>,</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">node.labels =</span> F, <span class="at">anc.states =</span> F))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="model-fitting-in-geomorph" class="level2">
<h2 class="anchored" data-anchor-id="model-fitting-in-geomorph">3.0 Model fitting in geomorph</h2>
<section id="defining-models" class="level3">
<h3 class="anchored" data-anchor-id="defining-models">3.1 Defining models</h3>
<p>For the final section of the tutorial, let’s clear up our workspace and load in the final example dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"larvalMorph"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s perform GPA.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>Y.gpa <span class="ot">&lt;-</span> <span class="fu">gpagen</span>(larvalMorph<span class="sc">$</span>tailcoords, <span class="at">curves =</span> larvalMorph<span class="sc">$</span>tail.sliders,</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">ProcD =</span> <span class="cn">FALSE</span>, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(Y.gpa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Next, let’s define some putative models of shape variation that would be biologically relevant.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>gdf <span class="ot">&lt;-</span> <span class="fu">geomorph.data.frame</span>(Y.gpa, <span class="at">treatment =</span> larvalMorph<span class="sc">$</span>treatment, </span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">family =</span> larvalMorph<span class="sc">$</span>family)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="co"># simple allometry model</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>fit.size <span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize), <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>) </span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co"># unique family allometries</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>fit.family<span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize) <span class="sc">*</span> family, <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>) </span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a><span class="co"># unique treatment: family allometries</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>fit.treatment<span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize) <span class="sc">*</span> treatment<span class="sc">/</span>family, <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s take a look at the models.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>fit.size</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Linear Model fit with lm.rrpp

Number of observations: 114
Number of dependent variables: 128  
Data space dimensions: 113  
Sums of Squares and Cross-products: Type I
Number of permutations: 1000
Call: procD.lm(f1 = coords ~ log(Csize), data = gdf, print.progress = FALSE)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>fit.family</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Linear Model fit with lm.rrpp

Number of observations: 114
Number of dependent variables: 128  
Data space dimensions: 113  
Sums of Squares and Cross-products: Type I
Number of permutations: 1000
Call: procD.lm(f1 = coords ~ log(Csize) * family, data = gdf, print.progress = FALSE)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>fit.treatment</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Linear Model fit with lm.rrpp

Number of observations: 114
Number of dependent variables: 128  
Data space dimensions: 113  
Sums of Squares and Cross-products: Type I
Number of permutations: 1000
Call: procD.lm(f1 = coords ~ log(Csize) * treatment/family, data = gdf,  
    print.progress = FALSE)</code></pre>
</div>
</div>
<p>Now we have defined our models, we can test how they fit our data using ANOVA.</p>
</section>
<section id="performing-anova-and-model-comparisons" class="level3">
<h3 class="anchored" data-anchor-id="performing-anova-and-model-comparisons">3.2 Performing ANOVA and model comparisons</h3>
<p>The following are the typical ANOVA tables for each model:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.size)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

            Df      SS      MS     Rsq      F      Z Pr(&gt;F)   
log(Csize)   1 0.50263 0.50263 0.16206 21.661 3.1141  0.001 **
Residuals  112 2.59885 0.02320 0.83794                        
Total      113 3.10147                                        
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = coords ~ log(Csize), data = gdf, print.progress = FALSE)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.family)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

                   Df      SS      MS     Rsq       F       Z Pr(&gt;F)   
log(Csize)          1 0.50263 0.50263 0.16206 22.5073  3.1333  0.001 **
family              3 0.21009 0.07003 0.06774  3.1360  2.2172  0.011 * 
log(Csize):family   3 0.02160 0.00720 0.00696  0.3224 -1.5886  0.952   
Residuals         106 2.36715 0.02233 0.76324                          
Total             113 3.10147                                          
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = coords ~ log(Csize) * family, data = gdf, print.progress = FALSE)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.treatment)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

                             Df      SS      MS     Rsq       F       Z Pr(&gt;F)
log(Csize)                    1 0.50263 0.50263 0.16206 26.6673  3.2716  0.001
treatment                     5 0.45807 0.09161 0.14770  4.8607  3.5710  0.001
log(Csize):treatment          5 0.08025 0.01605 0.02587  0.8515 -0.1177  0.553
log(Csize):treatment:family  18 0.47729 0.02652 0.15389  1.4068  1.3236  0.098
Residuals                    84 1.58323 0.01885 0.51048                       
Total                       113 3.10147                                       
                              
log(Csize)                  **
treatment                   **
log(Csize):treatment          
log(Csize):treatment:family . 
Residuals                     
Total                         
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = coords ~ log(Csize) * treatment/family, data = gdf,  
    print.progress = FALSE)</code></pre>
</div>
</div>
<p>We can compare any number of models at once.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.size, fit.family, fit.treatment, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Effect sizes (Z) based on F distributions

                                       ResDf Df    RSS      SS       MS     Rsq
coords ~ log(Csize) (Null)               112  1 2.5989                  0.00000
coords ~ log(Csize) * family             106  6 2.3672 0.23169 0.038616 0.07470
coords ~ log(Csize) * treatment/family    84 28 1.5832 1.01562 0.036272 0.32746
Total                                    113    3.1015                         
                                            F      Z     P Pr(&gt;F)
coords ~ log(Csize) (Null)                                       
coords ~ log(Csize) * family           1.7292 1.4262 0.083       
coords ~ log(Csize) * treatment/family 1.9244 2.7124 0.005       
Total                                                            </code></pre>
</div>
</div>
<p>One might ascertain that models that include parameters for family and treatment are significant improvements over a model that only contains size. Note that the model fits can use types I, II, or III, sums of squares and cross-products (SSCP), ordinary or generalized least squares, and the ANOVA can be performed on a number of different test statistics (see anova.lm.rrpp help file for more details).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>?anova.lm.rrpp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="allometry-analyses" class="level2">
<h2 class="anchored" data-anchor-id="allometry-analyses">4.0 Allometry analyses</h2>
<section id="simple-allometry-models" class="level3">
<h3 class="anchored" data-anchor-id="simple-allometry-models">4.1 Simple allometry models</h3>
<p>First, let’s just consider the simple allometry model:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit.size)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

            Df      SS      MS     Rsq      F      Z Pr(&gt;F)   
log(Csize)   1 0.50263 0.50263 0.16206 21.661 3.1141  0.001 **
Residuals  112 2.59885 0.02320 0.83794                        
Total      113 3.10147                                        
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = coords ~ log(Csize), data = gdf, print.progress = FALSE)</code></pre>
</div>
</div>
<p>It is clear that there is a significant association between shape and size. To visualize this relationship, there are many options. First, we can use the plot generic for procD.lm and vary some of the arguments. Let’s do that, using two different ways to visualize shape change: prediction lines (PredLine) and regression scores (RegScore). The former are first principal component scores for fitted values from the procD.lm fit (Adams and Nistri 2010); the latter are standardized projected shape scores, along the axis defined by the regression of shape on size (Drake and Klingenberg 2008).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(fit.size, <span class="at">type =</span> <span class="st">"regression"</span>, <span class="at">reg.type =</span> <span class="st">"PredLine"</span>, <span class="at">predictor =</span> <span class="fu">log</span>(gdf<span class="sc">$</span>Csize))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(fit.size, <span class="at">type =</span> <span class="st">"regression"</span>, <span class="at">reg.type =</span> <span class="st">"RegScore"</span>, <span class="at">predictor =</span> <span class="fu">log</span>(gdf<span class="sc">$</span>Csize))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The plot.procD.lm function is for any procD.lm fit. If one wishes to work specifically with allometry models, the plotAllometry function performs the same analysis as a convenient wrapper for plot.procD.lm. For example,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.size, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"PredLine"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.size, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"RegScore"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>An important detail with these plots is that PredLine and RegScore are model-based projections of shape data. As we will see below, changing the model changes the outcome of the plot.</p>
<p>We could also perform a two-block partial least squares (PLS) analysis to find the correlation between shape and size, which is not based on a particular model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>PLS <span class="ot">&lt;-</span> <span class="fu">two.b.pls</span>(<span class="fu">log</span>(gdf<span class="sc">$</span>Csize), gdf<span class="sc">$</span>coords, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>PLS</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
two.b.pls(A1 = log(gdf$Csize), A2 = gdf$coords, print.progress = FALSE) 



r-PLS: 0.466

Effect Size (Z): 3.1676

P-value: 0.001

Based on 1000 random permutations</code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PLS)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>An astute observer might catch that the PLS plot is exactly the same as the RegScore plot. It is, in this case of a simple allometry model. They are also both the same as a plot of the common allometric component (CAC, Mitteroecker et al.&nbsp;2004); i.e.,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.size, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"CAC"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The CAC plot will always be the same as the PLS plot, irrespective of the type of shape-allometry model. The RegScore plot is the same in this simple case because only one vector of regression coefficients is produced, which aligns perfectly with the major axis of covariation between shape and size (the CAC or the shape PLS vector; Adams et al.&nbsp;2013)</p>
<p>One can also append a size vector to a matrix of shape variables and perform principal components analysis (PCA), called size-shape PCA (Mitteroecker et al.&nbsp;2004).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.size, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"size.shape"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="complex-allometry-models" class="level3">
<h3 class="anchored" data-anchor-id="complex-allometry-models">4.2 Complex allometry models</h3>
<p>We already learned that family and treatment were “significant” model effects. Now let’s focus on whether we should believe that families or treatments have unique allometries or a common allometry.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>fit.unique <span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize) <span class="sc">*</span> treatment<span class="sc">/</span>family, </span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>) <span class="co"># unique allometries</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>fit.common <span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize) <span class="sc">+</span> treatment<span class="sc">/</span>family, </span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>) <span class="co"># common allometry</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.common, fit.unique, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Effect sizes (Z) based on F distributions

                                              ResDf Df    RSS       SS       MS
coords ~ log(Csize) + treatment/family (Null)    89  1 1.6587                  
coords ~ log(Csize) * treatment/family           84  5 1.5832 0.075471 0.015094
Total                                           113    3.1015                  
                                                   Rsq      F        Z     P
coords ~ log(Csize) + treatment/family (Null) 0.000000                      
coords ~ log(Csize) * treatment/family        0.024334 0.8008 -0.23384 0.598
Total                                                                       
                                              Pr(&gt;F)
coords ~ log(Csize) + treatment/family (Null)       
coords ~ log(Csize) * treatment/family              
Total                                               </code></pre>
</div>
</div>
<p>Because this model comparison did not yield a significant result, we can conclude that a common allometry model is appropriate. Thus, we might want to plot the results, color-coding the points by treatment</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.common, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"PredLine"</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">col =</span> <span class="fu">as.numeric</span>(gdf<span class="sc">$</span>treatment))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.common, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"RegScore"</span>,</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">col =</span> <span class="fu">as.numeric</span>(gdf<span class="sc">$</span>treatment))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The next section focuses on ANOVA for model effects and pairwise comparisons.</p>
</section>
</section>
<section id="updating-anova-and-performing-pairwise-comparisons" class="level2">
<h2 class="anchored" data-anchor-id="updating-anova-and-performing-pairwise-comparisons">5.0 Updating ANOVA and performing pairwise comparisons</h2>
<p>In our example, we have both fixed and random effects. Treatment is a fixed effect and family is a random effect, nested within our fixed effect (as egg clutches were “randomly”” sampled from the wild). Generally, when evaluating model effects, ANOVA involves assessing the probability of observed F-values which are ratios of mean squared (MS) values for effects to MS values for the appropriate random effect, usually the residuals. (For procD.lm models, the distribution of F-values is generated over many random permutations.) For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.common)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

                  Df      SS      MS     Rsq       F      Z Pr(&gt;F)   
log(Csize)         1 0.50263 0.50263 0.16206 26.9691 3.2807  0.001 **
treatment          5 0.45807 0.09161 0.14770  4.9157 3.5979  0.001 **
treatment:family  18 0.48207 0.02678 0.15543  1.4370 1.4376  0.075 . 
Residuals         89 1.65870 0.01864 0.53481                         
Total            113 3.10147                                         
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = coords ~ log(Csize) + treatment/family, data = gdf,  
    print.progress = FALSE)</code></pre>
</div>
</div>
<p>Notice that the F-value for the three effects - log(Csize), Treatment, and Treatment:Family - is calculated as MS effect / MS Residuals. This is the default. However, in our mixed-model ANOVA, we would prefer to calculate the F-value for treatment as MS Treatment / MS Treatment:Family, to determine if the treatment effect is meaningful compared to shape variation among families we sampled randomly and assigned to treatments. We can update our ANOVA by specifying what the denominator (error term) should be, as a sequence of error terms for the ANOVA; e.g.:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.common, <span class="at">error =</span> <span class="fu">c</span>(<span class="st">"Residuals"</span>, <span class="st">"treatment:family"</span>, <span class="st">"Residuals"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

                  Df      SS      MS     Rsq       F      Z Pr(&gt;F)   
log(Csize)         1 0.50263 0.50263 0.16206 26.9691 3.2807  0.001 **
treatment          5 0.45807 0.09161 0.14770  3.4208 1.9482  0.025 * 
treatment:family  18 0.48207 0.02678 0.15543  1.4370 1.4376  0.075 . 
Residuals         89 1.65870 0.01864 0.53481                         
Total            113 3.10147                                         
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = coords ~ log(Csize) + treatment/family, data = gdf,  
    print.progress = FALSE)</code></pre>
</div>
</div>
<p>Notice the F-value and effect size decreased a bit with recalculation of the F-value for treatment, but remained significant. What this function did was recalculate every treatment F-value in every random permutation to generate a new distribution for measuring effect size (Z) and P-value.</p>
<p>Now that we know that shape covaries with size, but in a common way for each treatment, we might wish to compare treatment least-squares (LS) means to see which treatments differ in shape, accounting for allometry and accounting for family effects. To achieve this, we will use the pairwise function from RRPP. There are several advantages to using the pairwise function. The most prominent are: (1) the ability to quickly change among different test types and summaries without re-analysis, (2) alternative summary options, (3) an option to use grouping variables not included in the original model fit, and (4) much faster computation for straightforward tests.</p>
<p>The pairwise function has this general format:</p>
<p>pairwise(fit, groups, covariate),</p>
<p>where fit is an already fitted object, using procD.lm, groups is a factor to designate groups to be compared, and covariate is a vector if slopes are to be compared. This format assumes the inherent null model of “fit” is appropriate. If an alternative null model is desired, the function can be updated as:</p>
<p>pairwise(fit, fit.null, groups, covariate),</p>
<p>where fit.null is a second procD.lm fit. If one is not sure about the inherent null model, they can use the reveal.model.designs function of RRPP to discover the exact null model used; e.g.,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">reveal.model.designs</span>(fit.common)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                 Reduced 
log(Csize)                             1 
treatment                     log(Csize) 
treatment:family  log(Csize) + treatment 
                                                      Full
log(Csize)                                      log(Csize)
treatment                           log(Csize) + treatment
treatment:family log(Csize) + treatment + treatment:family
                                                  
log(Csize)                                        
treatment                                         
treatment:family &lt;- Null/Full inherent in pairwise</code></pre>
</div>
</div>
<p>The results tell us that if we run pairwise on fit.common, the null model would be ~ log(Csize) + treatment and the full model would be ~ log(Csize) + treatment + treatment:family. This is the case because we used type I (sequential) sums of squares and cross-products (the default). However, it is maybe not ideal. We might prefer to have as a null model, ~ log(Csize) + family. Thus, let’s first establish that model and then run the pairwise function</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>fit.null <span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize) <span class="sc">+</span> family, <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>PW <span class="ot">&lt;-</span> <span class="fu">pairwise</span>(fit.common, fit.null, <span class="at">groups =</span> gdf<span class="sc">$</span>treatment, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>PW</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Pairwise comparisons

Groups: AG AM CF CI TG TM 

RRPP: 1000 permutations</code></pre>
</div>
</div>
<p>There are now many options for summarizing results; i.e., we can perform multiple tests! Here is one option:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PW, <span class="at">test.type =</span> <span class="st">"dist"</span>, <span class="at">confidence =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Pairwise comparisons

Groups: AG AM CF CI TG TM 

RRPP: 1000 permutations

LS means:
Vectors hidden (use show.vectors = TRUE to view)

Pairwise distances between means, plus statistics
               d  UCL (95%)          Z Pr &gt; d
AG:AM 0.03031748 0.08544041 -0.3983502  0.625
AG:CF 0.06775829 0.08787927  1.1355437  0.144
AG:CI 0.12685901 0.09140712  2.3133960  0.004
AG:TG 0.10993275 0.08320799  2.1315610  0.010
AG:TM 0.06447743 0.08823741  1.0581303  0.162
AM:CF 0.05742092 0.08919549  0.8048851  0.227
AM:CI 0.12904332 0.08977658  2.3407942  0.006
AM:TG 0.11418302 0.08878564  2.1302365  0.010
AM:TM 0.04829385 0.08787553  0.5319077  0.309
CF:CI 0.13108751 0.08447949  2.5457646  0.003
CF:TG 0.13207622 0.08781678  2.4905214  0.003
CF:TM 0.06377460 0.09022502  0.9788791  0.190
CI:TG 0.05102561 0.09150320  0.5093922  0.324
CI:TM 0.17096322 0.09526279  2.7628577  0.001
TG:TM 0.15667496 0.08594155  2.8008853  0.001</code></pre>
</div>
</div>
<p>The test statistics used, “dist”, is the distance between LS means. By specifying a confidence level, we are given upper confidence limits (UCL) from the distributions of pairwise distances. We can see that if the observed distance is larger than the UCL, the P-value is less than 1 - confidence; i.e., it is “significant”. The default is this “stats table”, but we could also produce pairwise tables. In fact, we can reproduce the old format for advanced.procD.lm like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.null, fit.common, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Effect sizes (Z) based on F distributions

                                       ResDf Df    RSS      SS       MS     Rsq
coords ~ log(Csize) + family (Null)      109  1 2.3887                  0.00000
coords ~ log(Csize) + treatment/family    89 20 1.6587 0.73005 0.036503 0.23539
Total                                    113    3.1015                         
                                            F     Z     P Pr(&gt;F)
coords ~ log(Csize) + family (Null)                             
coords ~ log(Csize) + treatment/family 1.9586 2.463 0.011       
Total                                                           </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PW, <span class="at">test.type =</span> <span class="st">"dist"</span>, <span class="at">confidence =</span> <span class="fl">0.95</span>, <span class="at">stat.table =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Pairwise comparisons

Groups: AG AM CF CI TG TM 

RRPP: 1000 permutations

LS means:
Vectors hidden (use show.vectors = TRUE to view)

Pairwise distances between means
           AG         AM         CF         CI         TG         TM
AG 0.00000000 0.03031748 0.06775829 0.12685901 0.10993275 0.06447743
AM 0.03031748 0.00000000 0.05742092 0.12904332 0.11418302 0.04829385
CF 0.06775829 0.05742092 0.00000000 0.13108751 0.13207622 0.06377460
CI 0.12685901 0.12904332 0.13108751 0.00000000 0.05102561 0.17096322
TG 0.10993275 0.11418302 0.13207622 0.05102561 0.00000000 0.15667496
TM 0.06447743 0.04829385 0.06377460 0.17096322 0.15667496 0.00000000

Pairwise 95% Upper confidence limits between means
           AG         AM         CF         CI         TG         TM
AG 0.00000000 0.08544041 0.08787927 0.09140712 0.08320799 0.08823741
AM 0.08544041 0.00000000 0.08919549 0.08977658 0.08878564 0.08787553
CF 0.08787927 0.08919549 0.00000000 0.08447949 0.08781678 0.09022502
CI 0.09140712 0.08977658 0.08447949 0.00000000 0.09150320 0.09526279
TG 0.08320799 0.08878564 0.08781678 0.09150320 0.00000000 0.08594155
TM 0.08823741 0.08787553 0.09022502 0.09526279 0.08594155 0.00000000

Pairwise effect sizes (Z) between means
           AG         AM        CF        CI        TG        TM
AG  0.0000000 -0.3983502 1.1355437 2.3133960 2.1315610 1.0581303
AM -0.3983502  0.0000000 0.8048851 2.3407942 2.1302365 0.5319077
CF  1.1355437  0.8048851 0.0000000 2.5457646 2.4905214 0.9788791
CI  2.3133960  2.3407942 2.5457646 0.0000000 0.5093922 2.7628577
TG  2.1315610  2.1302365 2.4905214 0.5093922 0.0000000 2.8008853
TM  1.0581303  0.5319077 0.9788791 2.7628577 2.8008853 0.0000000

Pairwise P-values between means
      AG    AM    CF    CI    TG    TM
AG 1.000 0.625 0.144 0.004 0.010 0.162
AM 0.625 1.000 0.227 0.006 0.010 0.309
CF 0.144 0.227 1.000 0.003 0.003 0.190
CI 0.004 0.006 0.003 1.000 0.324 0.001
TG 0.010 0.010 0.003 0.324 1.000 0.001
TM 0.162 0.309 0.190 0.001 0.001 1.000</code></pre>
</div>
</div>
<p>Because we have already performed the pairwise procedure, we could also summarize a different test. For example, let’s say we wish to compare morphological disparities (variances) among treatments. We simply change the summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PW, <span class="at">test.type =</span> <span class="st">"var"</span>, <span class="at">confidence =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Pairwise comparisons

Groups: AG AM CF CI TG TM 

RRPP: 1000 permutations


Observed variances by group

        AG         AM         CF         CI         TG         TM 
0.01419375 0.01384366 0.01923172 0.01704117 0.01006673 0.01287420 

Pairwise distances between variances, plus statistics
                 d   UCL (95%)           Z Pr &gt; d
AG:AM 0.0003500900 0.010087309 -1.69635391  0.949
AG:CF 0.0050379659 0.009938407  0.52168706  0.305
AG:CI 0.0028474188 0.010372592 -0.23118837  0.590
AG:TG 0.0041270191 0.010316444  0.22999925  0.433
AG:TM 0.0013195570 0.009850599 -0.87460939  0.789
AM:CF 0.0053880560 0.009613089  0.59377947  0.300
AM:CI 0.0031975089 0.010284072 -0.06700465  0.548
AM:TG 0.0037769290 0.009932049  0.19034887  0.443
AM:TM 0.0009694669 0.009679284 -1.08080047  0.850
CF:CI 0.0021905471 0.010327023 -0.48165241  0.674
CF:TG 0.0091649850 0.010154643  1.40254488  0.075
CF:TM 0.0063575229 0.010012406  0.85125709  0.211
CI:TG 0.0069744379 0.010679392  0.89506258  0.202
CI:TM 0.0041669758 0.010341725  0.26148124  0.410
TG:TM 0.0028074621 0.010485985 -0.20477972  0.578</code></pre>
</div>
</div>
<p>This should be exactly the same as performing a morphological disparity test</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">morphol.disparity</span>(fit.common, <span class="at">groups =</span> gdf<span class="sc">$</span>treatment, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
morphol.disparity(f1 = fit.common, groups = gdf$treatment, print.progress = FALSE) 




Randomized Residual Permutation Procedure Used
1000 Permutations

Procrustes variances for defined groups
        AG         AM         CF         CI         TG         TM 
0.01419375 0.01384366 0.01923172 0.01704117 0.01006673 0.01287420 


Pairwise absolute differences between variances
            AG           AM          CF          CI          TG           TM
AG 0.000000000 0.0003500900 0.005037966 0.002847419 0.004127019 0.0013195570
AM 0.000350090 0.0000000000 0.005388056 0.003197509 0.003776929 0.0009694669
CF 0.005037966 0.0053880560 0.000000000 0.002190547 0.009164985 0.0063575229
CI 0.002847419 0.0031975089 0.002190547 0.000000000 0.006974438 0.0041669758
TG 0.004127019 0.0037769290 0.009164985 0.006974438 0.000000000 0.0028074621
TM 0.001319557 0.0009694669 0.006357523 0.004166976 0.002807462 0.0000000000


P-Values
      AG    AM    CF    CI    TG    TM
AG 1.000 0.949 0.305 0.590 0.433 0.789
AM 0.949 1.000 0.300 0.548 0.443 0.850
CF 0.305 0.300 1.000 0.674 0.075 0.211
CI 0.590 0.548 0.674 1.000 0.202 0.410
TG 0.433 0.443 0.075 0.202 1.000 0.578
TM 0.789 0.850 0.211 0.410 0.578 1.000</code></pre>
</div>
</div>
<p>The pairwise function in RRPP is really versatile. More examples are provided in the help file for the function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>?pairwise</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Greater detail for how to summarize different tests is found in the summary.pairwise help file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>?summary.pairwise</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="performing-trajectory-analysis" class="level2">
<h2 class="anchored" data-anchor-id="performing-trajectory-analysis">6.0 Performing trajectory analysis</h2>
<p>The trajectory,analysis function has the same basic arguments as the pairwise function, but also has an argument for trajectory points (which can be a single value, if data are already trajectories or a factor to indicate trajectory point levels). Following the example above, trajectory analysis can be considered a pairwise function where treatments are trajectories and families are trajectory points. The following highlights the steps involved for one type of example (but the plotting options are quite numerous):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>TA <span class="ot">&lt;-</span> <span class="fu">trajectory.analysis</span>(fit.common, </span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">groups =</span> gdf<span class="sc">$</span>treatment, <span class="at">traj.pts =</span> gdf<span class="sc">$</span>family,</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">pca =</span> <span class="cn">TRUE</span>, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(TA, <span class="at">attribute =</span> <span class="st">"MD"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Trajectory analysis

1000 permutations.

Points projected onto trajectory PCs

Trajectories:
Trajectories hidden (use show.trajectories = TRUE to view)

Observed path distances by group

       AG        AM        CF        CI        TG        TM 
0.2122330 0.1916723 0.6845214 0.2411477 0.2356911 0.1719081 

Pairwise absolute differences in path distances, plus statistics
               d UCL (95%)          Z Pr &gt; d
AG:AM 0.02056069 0.2259968 -0.9720506  0.824
AG:CF 0.47228842 0.2316285  2.9185188  0.001
AG:CI 0.02891473 0.2373186 -0.7761383  0.779
AG:TG 0.02345814 0.2337741 -0.9994827  0.828
AG:TM 0.04032491 0.2185674 -0.5320552  0.702
AM:CF 0.49284911 0.2342246  2.9869742  0.001
AM:CI 0.04947542 0.2363919 -0.3521147  0.633
AM:TG 0.04401883 0.2309155 -0.4909832  0.685
AM:TM 0.01976423 0.2393851 -1.0673976  0.842
CF:CI 0.44337369 0.2338206  2.6970630  0.001
CF:TG 0.44883028 0.2483839  2.7829648  0.001
CF:TM 0.51261333 0.2365751  3.1039171  0.001
CI:TG 0.00545659 0.2305836 -1.8689261  0.969
CI:TM 0.06923965 0.2477913 -0.1034597  0.536
TG:TM 0.06378306 0.2455973 -0.0814343  0.537</code></pre>
</div>
</div>
<p>The argument, attribute = “MD”, indicates that the differences between trajectory magnitudes - the path length of trajectories - is considered. The trajectory analysis could also be summarized for trajectory correlations (TC), which are angles between trajectory directions (major axes of variation), or shape differences (SD). More examples are given in the trajectory.analysis help file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>?trajectory.analysis</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function, plot.trajectory.analysis plots the data points projected on the PCs for fitted values and the function, add.trajectories, superimposes the trajectories on these points.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>TP <span class="ot">&lt;-</span> <span class="fu">plot</span>(TA, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.7</span>, <span class="at">col =</span> <span class="fu">as.numeric</span>(gdf<span class="sc">$</span>treatment))</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="fu">add.trajectories</span>(TP, <span class="at">traj.bg =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nlevels</span>(gdf<span class="sc">$</span>treatment), </span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">start.bg =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nlevels</span>(gdf<span class="sc">$</span>treatment),</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">end.bg =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nlevels</span>(gdf<span class="sc">$</span>treatment))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>