<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Antigoni Kaliontzopoulou &amp; Michael Collyer. Adapted by Thomas Smith">

<title>An introduction to Geomorph</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="An_introduction_to_Geomorph_files/libs/clipboard/clipboard.min.js"></script>
<script src="An_introduction_to_Geomorph_files/libs/quarto-html/quarto.js"></script>
<script src="An_introduction_to_Geomorph_files/libs/quarto-html/popper.min.js"></script>
<script src="An_introduction_to_Geomorph_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="An_introduction_to_Geomorph_files/libs/quarto-html/anchor.min.js"></script>
<link href="An_introduction_to_Geomorph_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="An_introduction_to_Geomorph_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="An_introduction_to_Geomorph_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="An_introduction_to_Geomorph_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="An_introduction_to_Geomorph_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">An introduction to Geomorph</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Antigoni Kaliontzopoulou &amp; Michael Collyer. Adapted by Thomas Smith </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="preparing-your-data" class="level2">
<h2 class="anchored" data-anchor-id="preparing-your-data">1.0 Preparing your data</h2>
<p>In geomorph, you can digitize fixed landmarks, and semilandmarks on surfaces, using a 3D surface object or a mesh3d object. In its simplest version, this involves the following geomorph functions: digit.fixed, buildtemplate, and digitsurface. If you want to use curve semilandmarks, these may be either digitized as fixed landmarks, and then indexed for sliding using the curves argument of gpagen(); or sampled automatically from a digitized curve using digit.curves(). These functions require the package rgl, which in turn requires Java. As this requirement can cause problems for people (myself included) and is not the recommended approach for landmarking, I have not included it in this vignette.</p>
<p>First, let us check if you have geomorph installed, install it if necessary, then load the package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">c</span>(<span class="st">"geomorph"</span>)[<span class="sc">!</span><span class="fu">c</span>(<span class="st">"geomorph"</span>) <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[,<span class="st">"Package"</span>]]) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="fu">c</span>(<span class="st">"geomorph"</span>)[<span class="sc">!</span><span class="fu">c</span>(<span class="st">"geomorph"</span>) <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[,<span class="st">"Package"</span>]])</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geomorph)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: RRPP</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: rgl</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Matrix</code></pre>
</div>
</div>
</section>
<section id="principal-components-analysis-and-visualisation-tools-for-exploring-shape-space" class="level2">
<h2 class="anchored" data-anchor-id="principal-components-analysis-and-visualisation-tools-for-exploring-shape-space">2.0 Principal components analysis and visualisation tools for exploring shape space</h2>
<p>Version 3.1.0 of geomorph introduced the function gm.prcomp, and related utility functions (summary() and plot()), for performing principal components analyses on Procrustes shape variables for a set of aligned specimens. This function now includes several different types of analytical options and, combined with other visualization tools available in geomorph, provides tools for exploring variation in shape space.</p>
<p>Throughout, we will be using shape data of several Plethodon species as an example, so let´s first load and superimpose those.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list=</span><span class="fu">ls</span>())</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"plethspecies"</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>Y.gpa <span class="ot">&lt;-</span> <span class="fu">gpagen</span>(plethspecies<span class="sc">$</span>land, <span class="at">print.progress =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="traditional-pca-and-visualising-shape-patterns" class="level3">
<h3 class="anchored" data-anchor-id="traditional-pca-and-visualising-shape-patterns">2.1 Traditional PCA and visualising shape patterns</h3>
<p>One first option is to perform a “traditional” PCA, i.e.&nbsp;based on OLS-centering and projection of the data. This is very much like what is performed in the basic R function prcomp.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>PCA <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PCA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Principal Component Analysis 
Centering by OLS mean
Orthogonal projection of OLS residuals
Number of observations: 9 
Number of vectors 8 

Importance of Components:
                              Comp1        Comp2        Comp3        Comp4
Eigenvalues            0.0002720474 0.0001120524 0.0001084758 0.0000568924
Proportion of Variance 0.4564029477 0.1879858086 0.1819855633 0.0954461044
Cumulative Proportion  0.4564029477 0.6443887563 0.8263743196 0.9218204240
                              Comp5        Comp6        Comp7        Comp8
Eigenvalues            0.0000264508 1.260516e-05 5.959622e-06 1.584785e-06
Proportion of Variance 0.0443754550 2.114717e-02 9.998220e-03 2.658730e-03
Cumulative Proportion  0.9661958790 9.873431e-01 9.973413e-01 1.000000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCA, <span class="at">main =</span> <span class="st">"PCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Let’s customise this plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCA, <span class="at">main =</span> <span class="st">"PCA"</span>, <span class="at">pch =</span> <span class="dv">22</span>, <span class="at">bg =</span> <span class="st">"green"</span>, <span class="at">cex =</span> <span class="fl">1.5</span>, <span class="at">cex.lab =</span> <span class="fl">1.5</span>, <span class="at">font.lab =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>One then has several solutions for exploring shape variation across PC space and visualizing shape patterns. First, the user may choose to manually produce deformation grids to compare the shapes corresponding to the extremes of a chosen PC axis using plotRefToTarget. For example, one can:</p>
<ol type="1">
<li>Compare the minimum and maximum values to the global consensus:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>msh <span class="ot">&lt;-</span> <span class="fu">mshape</span>(Y.gpa<span class="sc">$</span>coords)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Minimum</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRefToTarget</span>(PCA<span class="sc">$</span>shapes<span class="sc">$</span>shapes.comp1<span class="sc">$</span>min, msh)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRefToTarget</span>(msh, PCA<span class="sc">$</span>shapes<span class="sc">$</span>shapes.comp1<span class="sc">$</span>max)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Maximum</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Comparing the minimum and maximum values:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRefToTarget</span>(PCA<span class="sc">$</span>shapes<span class="sc">$</span>shapes.comp1<span class="sc">$</span>min, PCA<span class="sc">$</span>shapes<span class="sc">$</span>shapes.comp1<span class="sc">$</span>max, <span class="at">method =</span> <span class="st">"vector"</span>, <span class="at">mag =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Of course here one can use all the plotting options available in plotRefToTarget. Please see the help file of that function for details.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>?plotRefToTarget</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="phylomorphospace" class="level3">
<h3 class="anchored" data-anchor-id="phylomorphospace">2.2 Phylomorphospace</h3>
<p>One may also want to project a phylogeny (if dealing with species-level observations), and estimated ancestral states into the ordination plot produced before, to obtain what is commonly referred to as a “phylomorphospace” plot. This can be easily done by providing a phylogenetic tree. This function estimates ancestral states before ordination takes place (i.e.&nbsp;using the raw data). For the Plethodon example data, we may project the phylogeny into the previous ordination plot as such:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>PCA.w.phylo <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy, <span class="at">GLS =</span> F)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PCA.w.phylo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Principal Component Analysis 
Centering by OLS mean
Orthogonal projection of OLS residuals
Number of observations: 9 
Number of vectors 8 

Importance of Components:
                              Comp1        Comp2        Comp3        Comp4
Eigenvalues            0.0002720474 0.0001120524 0.0001084758 0.0000568924
Proportion of Variance 0.4564029477 0.1879858086 0.1819855633 0.0954461044
Cumulative Proportion  0.4564029477 0.6443887563 0.8263743196 0.9218204240
                              Comp5        Comp6        Comp7        Comp8
Eigenvalues            0.0000264508 1.260516e-05 5.959622e-06 1.584785e-06
Proportion of Variance 0.0443754550 2.114717e-02 9.998220e-03 2.658730e-03
Cumulative Proportion  0.9661958790 9.873431e-01 9.973413e-01 1.000000e+00


Dispersion (variance) of points, after projection:
                                       Comp1        Comp2        Comp3
Tips Dispersion                 2.720474e-04 1.120524e-04 1.084758e-04
Proportion Tips Dispersion      4.564029e-01 1.879858e-01 1.819856e-01
Cumulative Tips Dispersion      4.564029e-01 6.443888e-01 8.263743e-01
Ancestors Dispersion            5.562296e-05 6.115844e-06 1.946905e-06
Proportion Ancestors Dispersion 7.523783e-01 8.272534e-02 2.633462e-02
Cumulative Ancestors Dispersion 7.523783e-01 8.351036e-01 8.614383e-01
                                       Comp4        Comp5        Comp6
Tips Dispersion                 5.689240e-05 2.645080e-05 1.260516e-05
Proportion Tips Dispersion      9.544610e-02 4.437545e-02 2.114717e-02
Cumulative Tips Dispersion      9.218204e-01 9.661959e-01 9.873431e-01
Ancestors Dispersion            7.984540e-06 1.524978e-06 2.962999e-07
Proportion Ancestors Dispersion 1.080021e-01 2.062746e-02 4.007870e-03
Cumulative Ancestors Dispersion 9.694403e-01 9.900678e-01 9.940757e-01
                                       Comp7        Comp8
Tips Dispersion                 5.959622e-06 1.584785e-06
Proportion Tips Dispersion      9.998220e-03 2.658730e-03
Cumulative Tips Dispersion      9.973413e-01 1.000000e+00
Ancestors Dispersion            1.945159e-07 2.434675e-07
Proportion Ancestors Dispersion 2.631099e-03 3.293238e-03
Cumulative Ancestors Dispersion 9.967068e-01 1.000000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCA.w.phylo, <span class="at">phylo =</span> T, <span class="at">main =</span> <span class="st">"PCA.w.phylo"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Note that the summary statistics obtained for this analysis are identical to those from the previous one. This is because here the phylogeny is merely used for plotting, and is NOT considered during the analytical procedures.</p>
<p>Again, all plotting arguments can be directly manipulated by the user. Please see the help file of plot.gm.prcomp for details.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>?plot.gm.prcomp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="phylopca" class="level3">
<h3 class="anchored" data-anchor-id="phylopca">2.3 phyloPCA</h3>
<p>Here, the phylogeny IS considered during the analytical step of the ordination, as the principal components analysis is in this case calculated based on GLS-centering and projection of the data. This is to account for the nonindependence among observations of species. For details on the analytical part of this method, see Revell 2009, Evolution 63: 3258 - 3268; Polly et al 2013, Hystrix 24: 33 - 41; Collyer &amp; Adams, submitted. This is effectively a PCA with a built in phylogenetic size correction.</p>
<p>For the Plethodon example data, this analysis would be implemented and plotted as follows (first with untransformed residual projection, second with transformed residual projection):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Phylo PCA without projecting untransformed residuals</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>phylo.PCA <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy, <span class="at">GLS =</span> <span class="cn">TRUE</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(phylo.PCA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Principal Component Analysis 
Centering by GLS mean
Oblique projection of GLS-centered residuals
Number of observations: 9 
Number of vectors 8 

Importance of Components:
                              Comp1        Comp2        Comp3        Comp4
Eigenvalues            2.761405e-05 9.322868e-06 7.889847e-06 6.605120e-06
Proportion of Variance 4.855705e-01 1.639351e-01 1.387365e-01 1.161457e-01
Cumulative Proportion  4.855705e-01 6.495056e-01 7.882421e-01 9.043878e-01
                              Comp5        Comp6        Comp7        Comp8
Eigenvalues            3.017886e-06 1.203779e-06 9.895110e-07 2.262231e-07
Proportion of Variance 5.306707e-02 2.116747e-02 1.739975e-02 3.977950e-03
Cumulative Proportion  9.574548e-01 9.786223e-01 9.960221e-01 1.000000e+00


Dispersion (variance) of points, after projection:
                                       Comp1        Comp2        Comp3
Tips Dispersion                 2.498092e-04 8.395185e-05 1.057114e-04
Proportion Tips Dispersion      4.190949e-01 1.408427e-01 1.773479e-01
Cumulative Tips Dispersion      4.190949e-01 5.599376e-01 7.372855e-01
Ancestors Dispersion            5.035969e-05 6.581394e-06 6.772535e-06
Proportion Ancestors Dispersion 6.811852e-01 8.902256e-02 9.160800e-02
Cumulative Ancestors Dispersion 6.811852e-01 7.702078e-01 8.618158e-01
                                       Comp4        Comp5        Comp6
Tips Dispersion                 1.027383e-04 2.765308e-05 1.675548e-05
Proportion Tips Dispersion      1.723599e-01 4.639247e-02 2.811000e-02
Cumulative Tips Dispersion      9.096453e-01 9.560378e-01 9.841478e-01
Ancestors Dispersion            8.668589e-06 7.070514e-07 3.112562e-07
Proportion Ancestors Dispersion 1.172548e-01 9.563859e-03 4.210175e-03
Cumulative Ancestors Dispersion 9.790705e-01 9.886344e-01 9.928446e-01
                                       Comp7        Comp8
Tips Dispersion                 7.598972e-06 1.850019e-06
Proportion Tips Dispersion      1.274849e-02 3.103703e-03
Cumulative Tips Dispersion      9.968963e-01 1.000000e+00
Ancestors Dispersion            2.791855e-07 2.498109e-07
Proportion Ancestors Dispersion 3.776375e-03 3.379042e-03
Cumulative Ancestors Dispersion 9.966210e-01 1.000000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(phylo.PCA, <span class="at">phylo =</span> <span class="cn">TRUE</span>, <span class="at">main =</span> <span class="st">"phylo PCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the transformed residuals</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>phylo.tPCA <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy, <span class="at">GLS =</span> <span class="cn">TRUE</span>, <span class="at">transform =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(phylo.tPCA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Principal Component Analysis 
Centering by GLS mean
GLS residuals transformed for orthogonal projection
Number of observations: 9 
Number of vectors 8 

Importance of Components:
                              Comp1        Comp2        Comp3        Comp4
Eigenvalues            2.761405e-05 9.322868e-06 7.889847e-06 6.605120e-06
Proportion of Variance 4.855705e-01 1.639351e-01 1.387365e-01 1.161457e-01
Cumulative Proportion  4.855705e-01 6.495056e-01 7.882421e-01 9.043878e-01
                              Comp5        Comp6        Comp7        Comp8
Eigenvalues            3.017886e-06 1.203779e-06 9.895110e-07 2.262231e-07
Proportion of Variance 5.306707e-02 2.116747e-02 1.739975e-02 3.977950e-03
Cumulative Proportion  9.574548e-01 9.786223e-01 9.960221e-01 1.000000e+00


Dispersion (variance) of points, after projection:
                                       Comp1        Comp2        Comp3
Tips Dispersion                 2.740155e-05 8.385080e-06 7.889441e-06
Proportion Tips Dispersion      4.933393e-01 1.509655e-01 1.420420e-01
Cumulative Tips Dispersion      4.933393e-01 6.443048e-01 7.863468e-01
Ancestors Dispersion            6.320992e-06 4.853117e-07 4.951270e-07
Proportion Ancestors Dispersion 7.646264e-01 5.870631e-02 5.989363e-02
Cumulative Ancestors Dispersion 7.646264e-01 8.233327e-01 8.832264e-01
                                       Comp4        Comp5        Comp6
Tips Dispersion                 6.590826e-06 2.909294e-06 1.175603e-06
Proportion Tips Dispersion      1.186617e-01 5.237912e-02 2.116563e-02
Cumulative Tips Dispersion      9.050085e-01 9.573876e-01 9.785532e-01
Ancestors Dispersion            8.025265e-07 6.382265e-08 1.498384e-08
Proportion Ancestors Dispersion 9.707858e-02 7.720384e-03 1.812539e-03
Cumulative Ancestors Dispersion 9.803049e-01 9.880253e-01 9.898379e-01
                                       Comp7        Comp8
Tips Dispersion                 9.878523e-07 2.033663e-07
Proportion Tips Dispersion      1.778536e-02 3.661419e-03
Cumulative Tips Dispersion      9.963386e-01 1.000000e+00
Ancestors Dispersion            6.135027e-08 2.265777e-08
Proportion Ancestors Dispersion 7.421309e-03 2.740825e-03
Cumulative Ancestors Dispersion 9.972592e-01 1.000000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(phylo.tPCA, <span class="at">phylo =</span> <span class="cn">TRUE</span>, <span class="at">main =</span> <span class="st">"phylo PCA with transformed projection"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>What looks different?</p>
</section>
<section id="paca-phylogenetically-aligned-pca" class="level3">
<h3 class="anchored" data-anchor-id="paca-phylogenetically-aligned-pca">2.4 PaCA: phylogenetically-aligned PCA</h3>
<p>This recently introduced method (Collyer &amp; Adams, submitted) provides an ordination that aligns phenotypic data with phylogenetic signal, by maximizing variation in directions that describe phylogenetic signal, while simultaneously preserving the Euclidean distances among observations in the data space. PaCA provides a projection that shows the most phylogenetic signal in the first few components, irrespective of other signals in the data. By comparing PCA, phyloPCA and PaCA results, one may glean the relative importance of phylogenetic and other (ecological) signals in the data.</p>
<p>For the Plethodon example data, this analysis would be implemented and plotted as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>PaCA <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy, <span class="at">align.to.phy =</span> <span class="cn">TRUE</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PaCA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Alignment to an alternative matrix 
Alignment matrix: phy 
Centering by OLS mean
OLS residuals
Alignment to  phy means residual projection is not orthogonal.
Number of observations: 9 
Number of vectors 8 

Importance of Components:
                                Comp1        Comp2        Comp3        Comp4
Singular Value           0.0003931629 0.0001155198 5.091258e-05 2.660786e-05
Proportion of Covariance 0.6595936040 0.1938028696 8.541399e-02 4.463894e-02
Cumulative Proportion    0.6595936040 0.8533964737 9.388105e-01 9.834494e-01
RV by Component          0.0993211764 0.0291827102 1.286158e-02 6.721702e-03
Cumulative RV            0.0993211764 0.1285038867 1.413655e-01 1.480872e-01
                                Comp5        Comp6        Comp7        Comp8
Singular Value           6.598675e-06 2.351469e-06 7.108871e-07 2.042572e-07
Proportion of Covariance 1.107033e-02 3.944966e-03 1.192627e-03 3.426742e-04
Cumulative Proportion    9.945197e-01 9.984647e-01 9.996573e-01 1.000000e+00
RV by Component          1.666963e-03 5.940304e-04 1.795850e-04 5.159965e-05
Cumulative RV            1.497541e-01 1.503482e-01 1.505278e-01 1.505794e-01


Dispersion (variance) of points, after projection:
                                       Comp1        Comp2        Comp3
Tips Dispersion                 2.301865e-04 1.124810e-04 1.133102e-04
Proportion Tips Dispersion      3.861747e-01 1.887049e-01 1.900960e-01
Cumulative Tips Dispersion      3.861747e-01 5.748795e-01 7.649755e-01
Ancestors Dispersion            3.820665e-05 9.237298e-06 9.698713e-06
Proportion Ancestors Dispersion 5.167983e-01 1.249474e-01 1.311886e-01
Cumulative Ancestors Dispersion 5.167983e-01 6.417457e-01 7.729343e-01
                                       Comp4        Comp5        Comp6
Tips Dispersion                 7.526878e-05 4.046474e-05 1.571235e-05
Proportion Tips Dispersion      1.262754e-01 6.788606e-02 2.635999e-02
Cumulative Tips Dispersion      8.912509e-01 9.591370e-01 9.854970e-01
Ancestors Dispersion            9.827479e-06 5.409542e-06 8.233072e-07
Proportion Ancestors Dispersion 1.329304e-01 7.317162e-02 1.113638e-02
Cumulative Ancestors Dispersion 9.058647e-01 9.790363e-01 9.901727e-01
                                       Comp7        Comp8
Tips Dispersion                 6.396433e-06 2.248360e-06
Proportion Tips Dispersion      1.073104e-02 3.771984e-03
Cumulative Tips Dispersion      9.962280e-01 1.000000e+00
Ancestors Dispersion            4.755226e-07 2.510041e-07
Proportion Ancestors Dispersion 6.432108e-03 3.395182e-03
Cumulative Ancestors Dispersion 9.966048e-01 1.000000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PaCA, <span class="at">phylo =</span> <span class="cn">TRUE</span>, <span class="at">main =</span> <span class="st">"PaCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="three-dimensional-pca-plot-with-a-phylogeny-and-time-on-the-z-axis" class="level3">
<h3 class="anchored" data-anchor-id="three-dimensional-pca-plot-with-a-phylogeny-and-time-on-the-z-axis">2.5 Three-dimensional PCA plot with a phylogeny and time on the z-axis</h3>
<p>Finally, plot.gm.prcomp provides the possibility of producing a 3D plot of any two PCA axes, with the phylogenetic tree connecting the observations and time on the z-axis. Again, different plotting parameters can be controlled to manipulate plot aesthetics. Note, that in this case an rgl plotting device will open for the 3D plot, but the corresponding biplot with the phylogeny projected (option 2, above) will also be produced. This function will output the 3D plot in a separate RGL window.</p>
<p>If this doesn’t work - don’t worry! You likely just need to install Java!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCA.w.phylo, <span class="at">time.plot =</span> <span class="cn">TRUE</span>, <span class="at">pch =</span> <span class="dv">22</span>, <span class="at">bg =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"red"</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="st">"green"</span>, <span class="dv">4</span>)), <span class="at">cex =</span> <span class="dv">2</span>, </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">phylo.par =</span> <span class="fu">list</span>(<span class="at">edge.color =</span> <span class="st">"grey60"</span>, <span class="at">edge.width =</span> <span class="fl">1.5</span>, <span class="at">tip.txt.cex =</span> <span class="fl">0.75</span>,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">node.labels =</span> F, <span class="at">anc.states =</span> F))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="model-fitting-in-geomorph" class="level2">
<h2 class="anchored" data-anchor-id="model-fitting-in-geomorph">3.0 Model fitting in geomorph</h2>
<section id="defining-models" class="level3">
<h3 class="anchored" data-anchor-id="defining-models">3.1 Defining models</h3>
<p>For the final section of the tutorial, let’s clear up our workspace and load in the final example dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"larvalMorph"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s perform GPA.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>Y.gpa <span class="ot">&lt;-</span> <span class="fu">gpagen</span>(larvalMorph<span class="sc">$</span>tailcoords, <span class="at">curves =</span> larvalMorph<span class="sc">$</span>tail.sliders,</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">ProcD =</span> <span class="cn">FALSE</span>, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(Y.gpa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Next, let’s define some putative models of shape variation that would be biologically relevant.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>gdf <span class="ot">&lt;-</span> <span class="fu">geomorph.data.frame</span>(Y.gpa, <span class="at">treatment =</span> larvalMorph<span class="sc">$</span>treatment, </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">family =</span> larvalMorph<span class="sc">$</span>family)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="co"># simple allometry model</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>fit.size <span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize), <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>) </span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="co"># unique family allometries</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>fit.family<span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize) <span class="sc">*</span> family, <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>) </span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="co"># unique treatment: family allometries</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>fit.treatment<span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize) <span class="sc">*</span> treatment<span class="sc">/</span>family, <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s take a look at the models.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>fit.size</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Linear Model fit with lm.rrpp

Number of observations: 114
Number of dependent variables: 128  
Data space dimensions: 113  
Sums of Squares and Cross-products: Type I
Number of permutations: 1000
Call: procD.lm(f1 = coords ~ log(Csize), data = gdf, print.progress = FALSE)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>fit.family</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Linear Model fit with lm.rrpp

Number of observations: 114
Number of dependent variables: 128  
Data space dimensions: 113  
Sums of Squares and Cross-products: Type I
Number of permutations: 1000
Call: procD.lm(f1 = coords ~ log(Csize) * family, data = gdf, print.progress = FALSE)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>fit.treatment</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Linear Model fit with lm.rrpp

Number of observations: 114
Number of dependent variables: 128  
Data space dimensions: 113  
Sums of Squares and Cross-products: Type I
Number of permutations: 1000
Call: procD.lm(f1 = coords ~ log(Csize) * treatment/family, data = gdf,  
    print.progress = FALSE)</code></pre>
</div>
</div>
<p>Now we have defined our models, we can test how they fit our data using ANOVA.</p>
</section>
<section id="performing-anova-and-model-comparisons" class="level3">
<h3 class="anchored" data-anchor-id="performing-anova-and-model-comparisons">3.2 Performing ANOVA and model comparisons</h3>
<p>The following are the typical ANOVA tables for each model:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.size)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

            Df      SS      MS     Rsq      F      Z Pr(&gt;F)    
log(Csize)   1 0.50263 0.50263 0.16206 21.661 3.1141  0.001 ***
Residuals  112 2.59885 0.02320 0.83794                         
Total      113 3.10147                                         
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = coords ~ log(Csize), data = gdf, print.progress = FALSE)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.family)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

                   Df      SS      MS     Rsq       F       Z Pr(&gt;F)    
log(Csize)          1 0.50263 0.50263 0.16206 22.5073  3.1333  0.001 ***
family              3 0.21009 0.07003 0.06774  3.1360  2.2172  0.011 *  
log(Csize):family   3 0.02160 0.00720 0.00696  0.3224 -1.5886  0.952    
Residuals         106 2.36715 0.02233 0.76324                           
Total             113 3.10147                                           
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = coords ~ log(Csize) * family, data = gdf, print.progress = FALSE)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.treatment)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

                             Df      SS      MS     Rsq       F       Z Pr(&gt;F)
log(Csize)                    1 0.50263 0.50263 0.16206 26.6673  3.2716  0.001
treatment                     5 0.45807 0.09161 0.14770  4.8607  3.5710  0.001
log(Csize):treatment          5 0.08025 0.01605 0.02587  0.8515 -0.1177  0.553
log(Csize):treatment:family  18 0.47729 0.02652 0.15389  1.4068  1.3236  0.098
Residuals                    84 1.58323 0.01885 0.51048                       
Total                       113 3.10147                                       
                               
log(Csize)                  ***
treatment                   ***
log(Csize):treatment           
log(Csize):treatment:family .  
Residuals                      
Total                          
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = coords ~ log(Csize) * treatment/family, data = gdf,  
    print.progress = FALSE)</code></pre>
</div>
</div>
<p>We can compare any number of models at once.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.size, fit.family, fit.treatment, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Effect sizes (Z) based on F distributions

                                       ResDf Df    RSS      SS       MS     Rsq
coords ~ log(Csize) (Null)               112  1 2.5989                  0.00000
coords ~ log(Csize) * family             106  6 2.3672 0.23169 0.038616 0.07470
coords ~ log(Csize) * treatment/family    84 28 1.5832 1.01562 0.036272 0.32746
Total                                    113    3.1015                         
                                            F      Z     P Pr(&gt;F)
coords ~ log(Csize) (Null)                                       
coords ~ log(Csize) * family           1.7292 1.4262 0.083       
coords ~ log(Csize) * treatment/family 1.9244 2.7124 0.005       
Total                                                            </code></pre>
</div>
</div>
<p>One might ascertain that models that include parameters for family and treatment are significant improvements over a model that only contains size. Note that the model fits can use types I, II, or III, sums of squares and cross-products (SSCP), ordinary or generalized least squares, and the ANOVA can be performed on a number of different test statistics (see anova.lm.rrpp help file for more details).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>?anova.lm.rrpp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="allometry-analyses" class="level2">
<h2 class="anchored" data-anchor-id="allometry-analyses">4.0 Allometry analyses</h2>
<p>It is important to note that I am no expert in the study of allometry, nor its analysis using geomorph. This is a wide-ranging package with a lot of different uses. What follows is a very brief foray into the analysis of allometry using geomorph for those who are interested using one of the models we defined earlier. There are much more detailed vignettes available online. I also highly recommend that those who are interested spend a few days reading through Chris Klingenberg’s work on the topic.</p>
<section id="simple-allometry-models" class="level3">
<h3 class="anchored" data-anchor-id="simple-allometry-models">4.1 Simple allometry models</h3>
<p>First, let’s just consider the simple allometry model we defined earlier:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit.size)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

            Df      SS      MS     Rsq      F      Z Pr(&gt;F)    
log(Csize)   1 0.50263 0.50263 0.16206 21.661 3.1141  0.001 ***
Residuals  112 2.59885 0.02320 0.83794                         
Total      113 3.10147                                         
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = coords ~ log(Csize), data = gdf, print.progress = FALSE)</code></pre>
</div>
</div>
<p>It is clear that there is a significant association between shape and size. To visualize this relationship, there are many options. First, we can use the plot generic for procD.lm and vary some of the arguments. Let’s do that, using two different ways to visualize shape change: prediction lines (PredLine) and regression scores (RegScore). The former are first principal component scores for fitted values from the procD.lm fit (Adams and Nistri 2010); the latter are standardized projected shape scores, along the axis defined by the regression of shape on size (Drake and Klingenberg 2008).</p>
<p>When interpreting these plots, keep in mind that the predictor is size and the other axis represents some axis of shape.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(fit.size, <span class="at">type =</span> <span class="st">"regression"</span>, <span class="at">reg.type =</span> <span class="st">"PredLine"</span>, <span class="at">predictor =</span> <span class="fu">log</span>(gdf<span class="sc">$</span>Csize))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(fit.size, <span class="at">type =</span> <span class="st">"regression"</span>, <span class="at">reg.type =</span> <span class="st">"RegScore"</span>, <span class="at">predictor =</span> <span class="fu">log</span>(gdf<span class="sc">$</span>Csize))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The plot.procD.lm function is for any procD.lm fit. If one wishes to work specifically with allometry models, the plotAllometry function performs the same analysis as a convenient wrapper for plot.procD.lm. You just specify the method using ‘method’ instead of ‘reg.type’. For example,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.size, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"PredLine"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.size, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"RegScore"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>An important detail with these plots is that PredLine and RegScore are model-based projections of shape data. As we will see below, changing the model changes the outcome of the plot.</p>
<p>For simple allometry models, RegScore is the same as the common allometric component (CAC, Mitteroecker et al.&nbsp;2004); i.e.,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.size, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"CAC"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The RegScore plot is the same in this simple case because only one vector of regression coefficients is produced, which aligns perfectly with the major axis of covariation between shape and size (the CAC; Adams et al.&nbsp;2013)</p>
<p>One can also append a size vector to a matrix of shape variables and perform principal components analysis (PCA), called size-shape PCA (Mitteroecker et al.&nbsp;2004).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.size, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"size.shape"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="complex-allometry-models" class="level3">
<h3 class="anchored" data-anchor-id="complex-allometry-models">4.2 Complex allometry models</h3>
<p>We already learned that family and treatment were “significant” model effects. Now let’s focus on whether we should believe that families or treatments have unique allometries or a common allometry.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>fit.unique <span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize) <span class="sc">*</span> treatment<span class="sc">/</span>family, </span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>) <span class="co"># unique allometries</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>fit.common <span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(coords <span class="sc">~</span> <span class="fu">log</span>(Csize) <span class="sc">+</span> treatment<span class="sc">/</span>family, </span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">data =</span> gdf, <span class="at">print.progress =</span> <span class="cn">FALSE</span>) <span class="co"># common allometry</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(fit.common, fit.unique, <span class="at">print.progress =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 1000 
Estimation method: Ordinary Least Squares 
Effect sizes (Z) based on F distributions

                                              ResDf Df    RSS       SS       MS
coords ~ log(Csize) + treatment/family (Null)    89  1 1.6587                  
coords ~ log(Csize) * treatment/family           84  5 1.5832 0.075471 0.015094
Total                                           113    3.1015                  
                                                   Rsq      F        Z     P
coords ~ log(Csize) + treatment/family (Null) 0.000000                      
coords ~ log(Csize) * treatment/family        0.024334 0.8008 -0.23384 0.598
Total                                                                       
                                              Pr(&gt;F)
coords ~ log(Csize) + treatment/family (Null)       
coords ~ log(Csize) * treatment/family              
Total                                               </code></pre>
</div>
</div>
<p>Because this model comparison did not yield a significant result, we can conclude that a common allometry model is appropriate. Thus, we might want to plot the results, color-coding the points by treatment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.common, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"PredLine"</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">col =</span> <span class="fu">as.numeric</span>(gdf<span class="sc">$</span>treatment))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllometry</span>(fit.common, <span class="at">size =</span> gdf<span class="sc">$</span>Csize, <span class="at">logsz =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"RegScore"</span>,</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">col =</span> <span class="fu">as.numeric</span>(gdf<span class="sc">$</span>treatment))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="testing-for-phylogenetic-signal-in-gpa-shape-variables" class="level2">
<h2 class="anchored" data-anchor-id="testing-for-phylogenetic-signal-in-gpa-shape-variables">5.0 Testing for phylogenetic signal in GPA shape variables</h2>
<p>Finally, geomorph gives us the ability to test for phylogenetic signal in our Procrustes shape variables without ordinating our data. Let’s load a new dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list=</span><span class="fu">ls</span>())</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(plethspecies)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Perform GPA and get shape variables.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>Y.gpa <span class="ot">&lt;-</span> <span class="fu">gpagen</span>(plethspecies<span class="sc">$</span>land)        </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Performing GPA

  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================================================| 100%

Making projections... Finished!</code></pre>
</div>
</div>
<p>Now to test for phylogenetic signal in shape. We will use physignal.z.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>PS.shape <span class="ot">&lt;-</span> <span class="fu">physignal.z</span>(<span class="at">A =</span> Y.gpa<span class="sc">$</span>coords, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy, <span class="at">lambda =</span> <span class="st">"front"</span>, <span class="at">iter =</span> <span class="dv">1000</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PS.shape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
physignal.z(A = Y.gpa$coords, phy = plethspecies$phy, lambda = "front",  
    iter = 1000) 

Evaluation of phylogenetic signal effect size
Optimization method: front
Optimization performed in 8 data dimensions.

Observed phylogenetic signal effect size (Z): -1.1739

P-value: 0.8127 based on 1001 random permutations

For a model with a log-likelihood of 277.4506
a branch-scaling (lambda) of 1e-04
and a ratio of Bownian Motion fit (K) of 0.8859 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PS.shape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PS.shape<span class="sc">$</span>PACA, <span class="at">phylo =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-37-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>There seems to be a roblem with an ill-conditioned residual covariance matrix. Let’s try shaving off one dimension (i.e.&nbsp;going from 8 dimensions to 7) by setting PAC.no to 7.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>PS.shape <span class="ot">&lt;-</span> <span class="fu">physignal.z</span>(<span class="at">A =</span> Y.gpa<span class="sc">$</span>coords, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy, <span class="at">lambda =</span> <span class="st">"front"</span>, <span class="at">PAC.no =</span> <span class="dv">7</span>, <span class="at">iter =</span> <span class="dv">1000</span>)</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PS.shape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
physignal.z(A = Y.gpa$coords, phy = plethspecies$phy, lambda = "front",  
    iter = 1000, PAC.no = 7) 

Evaluation of phylogenetic signal effect size
Optimization method: front
Optimization performed in 7 data dimensions.

Observed phylogenetic signal effect size (Z): 1.6933

P-value: 0.029 based on 1001 random permutations

For a model with a log-likelihood of 229.8667
a branch-scaling (lambda) of 1e-04
and a ratio of Bownian Motion fit (K) of 0.8901 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PS.shape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PS.shape<span class="sc">$</span>PACA, <span class="at">phylo =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-38-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now let’s test for phylogenetic signal in size.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>PS.size <span class="ot">&lt;-</span> <span class="fu">physignal.z</span>(<span class="at">A =</span> Y.gpa<span class="sc">$</span>Csize, <span class="at">phy =</span> plethspecies<span class="sc">$</span>phy, <span class="at">lambda =</span> <span class="st">"front"</span>, <span class="at">iter =</span> <span class="dv">1000</span>)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(PS.size)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
physignal.z(A = Y.gpa$Csize, phy = plethspecies$phy, lambda = "front",  
    iter = 1000) 

Evaluation of phylogenetic signal effect size
Optimization method: front
Optimization performed in 1 data dimensions.

Observed phylogenetic signal effect size (Z): -0.7885

P-value: 0.7712 based on 1001 random permutations

For a model with a log-likelihood of 66.0972
a branch-scaling (lambda) of 1e-04
and a ratio of Bownian Motion fit (K) of 0.6938 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PS.size)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="An_introduction_to_Geomorph_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Pagel’s lambda is the go-to measure of phylogenetic signal. Bounded between 0-1, estimated lambda values of 0 mean our traits are inferred to have no phylogenetic signal. 1 is Brownian motion (i.e.&nbsp;the only variation left after you take out the phylogenetic signal is random).</p>
<p>These functions also report Bloomberg’s K. Again, a value of 0 means independence, higher values = some sort of phylogenetic signal (it is unbounded).</p>
<p>The P-value reports significance of effect (i.e.&nbsp;low p-values = significant phylogenetic signal).</p>
<p>Finally, the functions also report a standardised effect size (Z).</p>
<p>Lots of information to parse but handily, one function does it all for you!</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>