<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Thomas Smith. Phytools sections by Liam Revell, adapted by Thomas Smith">

<title>Ordination and phylogenetic comparative methods</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Ordination_and_phylogenetic_comparative_methods_files/libs/clipboard/clipboard.min.js"></script>
<script src="Ordination_and_phylogenetic_comparative_methods_files/libs/quarto-html/quarto.js"></script>
<script src="Ordination_and_phylogenetic_comparative_methods_files/libs/quarto-html/popper.min.js"></script>
<script src="Ordination_and_phylogenetic_comparative_methods_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Ordination_and_phylogenetic_comparative_methods_files/libs/quarto-html/anchor.min.js"></script>
<link href="Ordination_and_phylogenetic_comparative_methods_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Ordination_and_phylogenetic_comparative_methods_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Ordination_and_phylogenetic_comparative_methods_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Ordination_and_phylogenetic_comparative_methods_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Ordination_and_phylogenetic_comparative_methods_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Ordination and phylogenetic comparative methods</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Thomas Smith. Phytools sections by Liam Revell, adapted by Thomas Smith </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">1.0 Introduction</h2>
<p>This tutorial will introduce you to the various functions you can use to ordinate your data outside of the Vegan package. It will also take you show you how to perform stochastic character mapping (a form of ancestral state estimation) using Phytools and how to test for phylogenetic signal in categorical and continuous data using the Geiger package.</p>
<section id="load-packages" class="level3">
<h3 class="anchored" data-anchor-id="load-packages">1.1 Load packages</h3>
<p>We will generate distance matrices using base R, Vegan, dispRity, and Claddis. Let’s make sure the latter packages are installed and load them.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">c</span>(<span class="st">"phytools"</span>, <span class="st">"geiger"</span>, <span class="st">"Claddis"</span>, <span class="st">"dispRity"</span>, <span class="st">"caper"</span>)[<span class="sc">!</span><span class="fu">c</span>(<span class="st">"phytools"</span>, <span class="st">"geiger"</span>, <span class="st">"Claddis"</span>, <span class="st">"dispRity"</span>, <span class="st">"caper"</span>) <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[,<span class="st">"Package"</span>]]) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="fu">c</span>(<span class="st">"phytools"</span>, <span class="st">"geiger"</span>, <span class="st">"Claddis"</span>, <span class="st">"dispRity"</span>, <span class="st">"caper"</span>)[<span class="sc">!</span><span class="fu">c</span>(<span class="st">"phytools"</span>, <span class="st">"geiger"</span>, <span class="st">"Claddis"</span>, <span class="st">"dispRity"</span>, <span class="st">"caper"</span>) <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[,<span class="st">"Package"</span>]])</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="principal-components-analysis-by-hand" class="level3">
<h3 class="anchored" data-anchor-id="principal-components-analysis-by-hand">1.2 Principal components analysis by hand</h3>
<p>This may seem exhaustive but I think it’s useful to break these analyses down. Let’s generate a data matrix first.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">sample</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dv">100</span>, <span class="at">replace =</span> T), <span class="at">nrow =</span> <span class="dv">10</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(data) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"Species_"</span>,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">10</span>,<span class="dv">1</span>))</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(data) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"C"</span>,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">10</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s get the column means so we can can see how we center our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>col.means <span class="ot">&lt;-</span> <span class="fu">apply</span>(data, <span class="dv">2</span>, mean)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># get centred matrix</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>centred.matrix <span class="ot">&lt;-</span> data</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>centred.matrix[,<span class="dv">1</span>] <span class="ot">&lt;-</span> centred.matrix[,<span class="dv">1</span>]<span class="sc">-</span>col.means[<span class="dv">1</span>]</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>centred.matrix[,<span class="dv">2</span>] <span class="ot">&lt;-</span> centred.matrix[,<span class="dv">2</span>]<span class="sc">-</span>col.means[<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Fortunately, the cov function does this step for us before computing a covariance matrix. Let’s do that now.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>covariance <span class="ot">&lt;-</span> <span class="fu">cov</span>(data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, let’s derive our eigenvectors and eigenvalues.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>eigenvalues <span class="ot">&lt;-</span> <span class="fu">eigen</span>(covariance)<span class="sc">$</span>values</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>eigenvectors <span class="ot">&lt;-</span> <span class="fu">eigen</span>(covariance)<span class="sc">$</span>vectors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Almost there! We have our axes! To get the scores for our ‘species’, we need multiple our centred matrix by our eigenvector matrix (aren’t you glad we have functions to do this for us?)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get principal component scores</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># we do this by getting the dot product</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>PCs <span class="ot">&lt;-</span> centred.matrix<span class="sc">%*%</span>eigenvectors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s plot our ordination.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="base-r" class="level2">
<h2 class="anchored" data-anchor-id="base-r">2.0 Base R</h2>
<section id="principal-components-analysis" class="level3">
<h3 class="anchored" data-anchor-id="principal-components-analysis">2.1 Principal components analysis</h3>
<p>Let’s start simple with base R. Ordination in base R is wonderfully simple. For principal component analysis, we can use the prcomp function. First, let’s generate a continuous dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>c.data <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">runif</span>(<span class="dv">100</span>,<span class="dv">0</span>,<span class="dv">10</span>), <span class="at">nrow =</span> <span class="dv">5</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(c.data) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"Species_"</span>,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">5</span>,<span class="dv">1</span>))</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(c.data) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"C"</span>,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">20</span>,<span class="dv">1</span>))</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(c.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s ordinate it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>PCA <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(c.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This doesn’t provide the eigenvalues by default, so we’ll need to use the eigen function to acquire them (there is a unique solution to eigen-based methods, so these will match the eigenvectors of our PCA object).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>c.eigen <span class="ot">&lt;-</span> <span class="fu">eigen</span>(<span class="fu">cov</span>(c.data), <span class="at">only.values =</span> T)<span class="sc">$</span>values</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s convert these into relative eigenvalues so we can so how the total variance is apportioned across the different axes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(c.eigen) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC"</span>,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">20</span>,<span class="dv">1</span>))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>c.eigen <span class="ot">&lt;-</span> c.eigen<span class="sc">/</span><span class="fu">sum</span>(c.eigen)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(c.eigen, <span class="at">ylab =</span> <span class="st">"Proportion of eigenvalues"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Looks like the first four axes are by far the most informative, with the first two capturing over 50% of the variance together. Let’s plot these for now.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCA<span class="sc">$</span>x, <span class="at">xlab =</span> <span class="fu">paste0</span>(<span class="st">"PC1 ("</span>,c.eigen[<span class="dv">1</span>]<span class="sc">*</span><span class="dv">100</span>,<span class="st">"% variance)"</span>), <span class="at">ylab =</span> <span class="fu">paste0</span>(<span class="st">"PC2 ("</span>,c.eigen[<span class="dv">2</span>]<span class="sc">*</span><span class="dv">100</span>,<span class="st">"% variance)"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="principal-coordinates-analysis" class="level3">
<h3 class="anchored" data-anchor-id="principal-coordinates-analysis">2.2 Principal coordinates analysis</h3>
<p>Lets generate a Euclidean distance matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s ordinate it using cmdscale (the base R function for principal coordinates analysis). Now it is worth noting that by default that cmdscale fits the data to just two axes. We don’t want this. The maximum is n-1, where n is the number of points/species. In this case, we have 10, so let’s set k to 9.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>PCoA <span class="ot">&lt;-</span> <span class="fu">cmdscale</span>(<span class="fu">sqrt</span>(dist), <span class="at">k =</span> <span class="dv">9</span>, <span class="at">eig =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s check the eigenvalues. Cmdscale returns them for us if eigen = TRUE.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>d.eigen <span class="ot">&lt;-</span> PCoA<span class="sc">$</span>eig</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(d.eigen) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC"</span>,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">10</span>,<span class="dv">1</span>))</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>d.eigen <span class="ot">&lt;-</span> d.eigen<span class="sc">/</span><span class="fu">sum</span>(d.eigen)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(d.eigen, <span class="at">ylab =</span> <span class="st">"Proportion of eigenvalues"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The variation hasn’t been summarised particularly well here, so any visualisations we derive will be limited in utility. We need to take a more quantitative approach to analysing this data (more on this soon!)</p>
</section>
</section>
<section id="ape" class="level2">
<h2 class="anchored" data-anchor-id="ape">3.0 Ape</h2>
<p>Ape provides an alternative to cmdscale that allows you to apply the Cailliez or Lingoes correction during the ordination process. This is done through the correction argument. It also returns all of the eigenvalues and eigenvectors by default, EXCEPT for those with negative eigenvalues. As such, you have to transform your data and/or correct it each time you use this function to be sure you are not working with a misrepresentation of your data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># square root transformation</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>ape.PCoA.sq <span class="ot">&lt;-</span> <span class="fu">pcoa</span>(<span class="fu">sqrt</span>(dist), <span class="at">correction =</span> <span class="st">"none"</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co"># square root transformation + Cailliez correction</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>ape.PCoA.C <span class="ot">&lt;-</span> <span class="fu">pcoa</span>(<span class="fu">sqrt</span>(dist), <span class="at">correction =</span> <span class="st">"cailliez"</span>)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="co"># square root transformation + Lingoes correction</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>ape.PCoA.L <span class="ot">&lt;-</span> <span class="fu">pcoa</span>(<span class="fu">sqrt</span>(dist), <span class="at">correction =</span> <span class="st">"lingoes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Why not try plotting them and seeing how they differ? What about from an ordination of untransformed, uncorrected data?</p>
</section>
<section id="claddis" class="level2">
<h2 class="anchored" data-anchor-id="claddis">4.0 Claddis</h2>
<p>Let’s try ordinating data in Claddis! First let’s load the library and load the dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Claddis)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: phytools</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: maps</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: strap</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: geoscale</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(day_2016)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s drop the three continuous characters as before.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>day_2016[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Claddis only supports principal coordinates analysis. However it does this all in one.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>clad.PCoA <span class="ot">&lt;-</span> <span class="fu">ordinate_cladistic_matrix</span>(day_2016, <span class="at">distance_metric =</span> <span class="st">"mord"</span>, <span class="at">distance_transformation =</span> <span class="st">"sqrt"</span>, <span class="at">correction =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The following taxa had to be removed to produce a complete distance matrix: Lycaenodon_longiceps</code></pre>
</div>
</div>
<p>You’ll notice we received a warning that Claddis was unable to produce a complete distance matrix (essential for ordination). We can do this ahead of time by submitting a distance matrix to the trim_matrix function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>clad.dist <span class="ot">&lt;-</span> <span class="fu">calculate_morphological_distances</span>(day_2016, <span class="at">distance_metric =</span> <span class="st">"mord"</span>, <span class="at">distance_transformation =</span> <span class="st">"sqrt"</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>trimmed.clad.dist <span class="ot">&lt;-</span> <span class="fu">trim_matrix</span>(clad.dist<span class="sc">$</span>distance_matrix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can see the taxa that have been trimmed by accessing the “removed_taxa” element of the object produced by trim_matrix. You can isolate the new distance matrix and add it to the distance matrix object as well.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>trimmed.clad.dist<span class="sc">$</span>removed_taxa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Lycaenodon_longiceps"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>clad.dist<span class="sc">$</span>distance_matrix <span class="ot">&lt;-</span> trimmed.clad.dist<span class="sc">$</span>distance_matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Anyway, back to the ordination object. Claddis calls the ape pcoa function to handle its ordination. As such, it can apply the Cailliez and Lingoes correction through the correction argument, as well as apply a square root transformation. Unfortunately, the ape pcoa function does not return the eigenvectors and eigenvalues of any negative eigenvalues that might result, so we can’t check if we need to transform our data in the first place. However, no information is lost during the square root transformation so it doesn’t hurt to apply it in all cases.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(clad.PCoA<span class="sc">$</span>values<span class="sc">$</span>Relative_eig, <span class="at">ylab =</span> <span class="st">"Proportion of eigenvalues"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Claddis has a wide variety of plotting functions that allow you to generate two and three dimensional plots. The standard, plot_morphospace, allows you to plot regular plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_morphospace</span>(clad.PCoA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It allows you to add convex hulls.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define groups</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>taxon_groups <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">nonBurnetiamorpha =</span> <span class="fu">c</span>(<span class="st">"Biarmosuchus_tener"</span>,</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="st">"Hipposaurus_boonstrai"</span>, <span class="st">"Bullacephalus_jacksoni"</span>, <span class="st">"Pachydectes_elsi"</span>,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="st">"Ictidorhinus_martinsi"</span>, <span class="st">"RC_20"</span>, <span class="st">"Herpetoskylax_hopsoni"</span>),</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="at">Burnetiamorpha =</span> <span class="fu">c</span>(<span class="st">"Lemurosaurus_pricei"</span>, <span class="st">"Lobalopex_mordax"</span>,</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="st">"Lophorhinus_willodenensis"</span>, <span class="st">"Proburnetia_viatkensis"</span>, <span class="st">"Lende_chiweta"</span>,</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="st">"Paraburnetia_sneeubergensis"</span>, <span class="st">"Burnetia_mirabilis"</span>, <span class="st">"BP_1_7098"</span>))</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(taxon_groups) <span class="ot">&lt;-</span> <span class="st">"taxonGroups"</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="co"># plot</span></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_morphospace</span>(<span class="at">pcoa_input =</span> clad.PCoA, <span class="at">plot_taxon_names =</span> F, <span class="at">taxon_groups =</span> taxon_groups, <span class="at">plot_convex_hull =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It allows you to add trees.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define tree - you can load in a nexus file or .tre file instead of writing it manually.</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>time_tree <span class="ot">&lt;-</span> <span class="fu">read.tree</span>(<span class="at">text =</span> <span class="fu">paste0</span>(<span class="st">"(Biarmosuchus_tener:0.5,"</span>,</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="st">"(((Hipposaurus_boonstrai:3.5,(Bullacephalus_jacksoni:0.75,"</span>,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="st">"Pachydectes_elsi:0.75):0.75):0.75,(Lemurosaurus_pricei:7.166666667,"</span>,</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="st">"(Lobalopex_mordax:4.333333333,((Lophorhinus_willodenensis:3.666666667,"</span>,</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="st">"(Proburnetia_viatkensis:0.8333333333,(Lende_chiweta:2,"</span>,</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="st">"(Paraburnetia_sneeubergensis:1,Burnetia_mirabilis:2):1):1.833333333)"</span>,</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="st">":0.8333333333):0.8333333333,(BP_1_7098:2.25,Niuksenitia_sukhonensis:"</span>,</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="st">"1.25):1.25):0.8333333333):0.8333333333):3.083333333):1.95,"</span>,</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a><span class="st">"(Ictidorhinus_martinsi:15.9,(RC_20:11.6,(Herpetoskylax_hopsoni:11.3,"</span>,</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a><span class="st">"Lycaenodon_longiceps:0.3):0.3):0.3):0.3):0.3);"</span>))</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a><span class="co"># add a root.time element, specifying just that</span></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>time_tree<span class="sc">$</span>root.time <span class="ot">&lt;-</span> <span class="fl">269.5</span></span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a><span class="co"># now run the ordination again, submitting the tree via argument time_tree</span></span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>clad.PCoA <span class="ot">&lt;-</span> <span class="fu">ordinate_cladistic_matrix</span>(day_2016, <span class="at">distance_transformation =</span> <span class="st">"sqrt"</span>, <span class="at">correction =</span> <span class="st">"none"</span>, <span class="at">time_tree =</span> time_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').


Note:
   This function is redundant with 'phytools::ancr' in situations in
   which it should be used (symmetric Q matrices) &amp; invalid for non-
   symmetric Q matrices (e.g., model='ARD').</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following taxa or nodes had to be removed to produce a complete distance matrix: BP_1_7098%%Niuksenitia_sukhonensis, Niuksenitia_sukhonensis</code></pre>
</div>
</div>
<p>You can see a couple of taxa have been dropped for the sake of a complete distance matrix automatically. The warnings that pop up are a reflection that Claddis is slightly out of date. As such, you’ll probably want to use other methods for generating your ancestral state data and then just derive a distance matrix from a data matrix including tip and node values, ordinate it without submitting a tree, and then insert it into an object that was generating using a tree to replace the dubious PCoA scores (this is why Claddis is difficult). Anyway, let’s plot the result:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_morphospace</span>(clad.PCoA, <span class="at">plot_taxon_names =</span> F, <span class="at">taxon_groups =</span> taxon_groups, <span class="at">plot_convex_hulls =</span> <span class="cn">TRUE</span>, <span class="at">plot_internal_nodes =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>See the below help documentation to explore what the other functions can do for you:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>?plot_chronophylomorphospace</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>?plot_morphospace_stack</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>?plot_multi_morphospace</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="disprity" class="level2">
<h2 class="anchored" data-anchor-id="disprity">5.0 dispRity</h2>
<p>dispRity has functionality for importing data from Claddis and geomorph and ordinating it, using principal coordinates analysis and principal components analysis respectively. Let’s take a look.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dispRity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="claddis-ordination" class="level3">
<h3 class="anchored" data-anchor-id="claddis-ordination">5.1 Claddis ordination</h3>
<p>Now let’s load in some Claddis data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"michaux_1989"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we ordinate!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>dispR.clad.ord <span class="ot">&lt;-</span> <span class="fu">Claddis.ordination</span>(michaux_1989, <span class="at">distance =</span> <span class="st">"mord"</span>, <span class="at">distance_transformation =</span> <span class="st">"sqrt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>dispRity doesn’t have its own plotting function. However, that just means you can use base R to plot it. There are other packages we will take a look at later on during the course.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(dispR.clad.ord)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You might have noticed the output of this function is pretty bare. This is because it passes the ordination off to cmdscale (base R function for PCoA). You should be able to pass on a request to this function to return the eigenvalues using the arg.cmdscale argument of the Claddis.ordination function, but this doesn’t seem to work consistently (might be bugged).</p>
</section>
<section id="geomorph-ordination" class="level3">
<h3 class="anchored" data-anchor-id="geomorph-ordination">5.2 Geomorph ordination</h3>
<p>Let’s load in some data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geomorph)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: RRPP</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'RRPP'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dispRity':

    add.tree</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: rgl</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'geomorph'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dispRity':

    combine.subsets</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"plethodon"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s ordinate it (after we perform GPS).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>procrustes <span class="ot">&lt;-</span> <span class="fu">gpagen</span>(plethodon<span class="sc">$</span>land)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Performing GPA

  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================================================| 100%

Making projections... Finished!</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>dispR.geo.ord <span class="ot">&lt;-</span> <span class="fu">geomorph.ordination</span>(procrustes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can also ordinate a geomorph data frame (handy for compatability).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>geomorph_df <span class="ot">&lt;-</span> <span class="fu">geomorph.data.frame</span>(procrustes, <span class="at">species =</span> plethodon<span class="sc">$</span>species)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geomorph.ordination</span>(geomorph_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ---- dispRity object ---- 
2 customised subsets for 40 elements in one matrix:
    species.Jord, species.Teyah.</code></pre>
</div>
</div>
</section>
</section>
<section id="testing-for-phylogenetic-signal-in-discrete-and-continuous-non-landmark-data-using-geiger-and-phytools" class="level2">
<h2 class="anchored" data-anchor-id="testing-for-phylogenetic-signal-in-discrete-and-continuous-non-landmark-data-using-geiger-and-phytools">6.0 Testing for phylogenetic signal in discrete and continuous (non-landmark) data using Geiger and Phytools</h2>
<p>Pagel’s lambda is a robust measure of the amount of non-random variance in a dataset is explained by the associated tree. The R package phytools let’s us derive an estimate of lambda by fitting a model of phylogenetic signal to our data. This works for continuous data. Geiger let’s us derive it for discrete data. The former is particularly important, as this means you can test for phylogenetic in any ordinated data set (PC scores are continuous).</p>
<p>Load the library. We’ll load phytools too as it will give us some data to work with.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phytools)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="continuous-data" class="level3">
<h3 class="anchored" data-anchor-id="continuous-data">6.1 Continuous data</h3>
<p>Load the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"anoletree"</span>)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"anole.data"</span>)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="fu">name.check</span>(anoletree, anole.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "OK"</code></pre>
</div>
</div>
<p>Using phytools, you can calculate Bloomberg’s K, which compares the variance of PICs to what we would espect under a Brownian motion model. K = 1 means that relatives resemble one another as much as we should expect under BM; K &lt; 1 means that there is less “phylogenetic signal” than expected under BM, while K &gt; 1 means that there is more. A significant p-value returned from phylosignal tells you that there is significant phylogenetic signal - that is, close relatives are more similar than random pairs of species.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>anoleSize <span class="ot">&lt;-</span> anole.data[, <span class="dv">1</span>]</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(anoleSize) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(anole.data)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">phylosig</span>(anoletree, anoleSize, <span class="at">test =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic signal K : 1.69526 
P-value (based on 1000 randomizations) : 0.001 </code></pre>
</div>
</div>
<p>Now let’s calculate Pagel’s lambda. Lambda is a tree transformation that stretches tip branches relative to internal branches, making the tree more and more like a complete polytomy. If our estimated lambda = 0, then the traits are inferred to have no phylogenetic signal. Lambda = 1 corresponds to a Brownian motion model; 0 &lt; lambda &lt; 1 is in between.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">phylosig</span>(anoletree, anoleSize, <span class="at">method =</span> <span class="st">"lambda"</span>, <span class="at">test =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic signal lambda : 0.999927 
logL(lambda) : 5.25315 
LR(lambda=0) : 97.8975 
P-value (based on LR test) : 4.40613e-23 </code></pre>
</div>
</div>
<p>That is some strong phylogenetic signal! You could also achieve this by using fitContinuous in Geiger. Let’s compare this to a simple Brownian motion model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>lambdaModel <span class="ot">&lt;-</span> <span class="fu">fitContinuous</span>(anoletree, anoleSize, <span class="at">model =</span> <span class="st">"lambda"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in fitContinuous(anoletree, anoleSize, model = "lambda"): 
Parameter estimates appear at bounds:
    lambda</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>brownianModel <span class="ot">&lt;-</span> <span class="fu">fitContinuous</span>(anoletree, anoleSize)</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>lambdaModel<span class="sc">$</span>opt<span class="sc">$</span>aicc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -4.204549</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>brownianModel<span class="sc">$</span>opt<span class="sc">$</span>aicc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -6.360343</code></pre>
</div>
</div>
<p>The lower the better when it comes to AIC scores, so our lambda model is a better fit for the data. You can fit all sorts of models with fitContinuous. Check out the documentation and have a think about whether you might want fit any others.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>?fitContinuous</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="discrete-data." class="level3">
<h3 class="anchored" data-anchor-id="discrete-data.">6.2 Discrete data.</h3>
<p>With discrete data, we need to use a different statistic, Fritz &amp; Purvis’s D. It is the sum of changes in estimated nodal values of a binary trait along edges in a phylogeny (D) provides a measure of the phylogenetic signal in that trait (Fritz and Purvis, 2010). If a trait is highly conserved, with only a basal division between two clades expressing either trait value, then the only change will be along the two daughters at the root. This will give a summed value of 1: the two differences between the root nodal value of 0.5 and the ancestors of the 1 and 0 clades. In contrast, if the trait is labile, more differences will be observed and the sum will be higher.</p>
<p>We will calculate this using the phylo.D function from the caper package. First we will load the data and convert it to the necessay format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caper)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: MASS</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: mvtnorm</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(BritishBirds)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>BritishBirds <span class="ot">&lt;-</span> <span class="fu">comparative.data</span>(BritishBirds.tree, BritishBirds.data, binomial)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Calculate Fritz &amp; Purvis’ D</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>redPhyloD <span class="ot">&lt;-</span> <span class="fu">phylo.d</span>(BritishBirds, <span class="at">binvar=</span>Red_list)</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(redPhyloD)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Calculation of D statistic for the phylogenetic structure of a binary variable

  Data :  BritishBirds.data
  Binary variable :  Red_list
  Counts of states:  0 = 149
                     1 = 32
  Phylogeny :  BritishBirds.tree
  Number of permutations :  1000

Estimated D :  0.5597442
Probability of E(D) resulting from no (random) phylogenetic structure :  0.001
Probability of E(D) resulting from Brownian phylogenetic structure    :  0.017</code></pre>
</div>
</div>
<p>The estimated D is below 1 and the probablility of this pattern emerging from Brownian phylogenetic structure is higher than it being random. This indicates that phylogenetic signal is at least partially responsible or the variance in the data.</p>
</section>
</section>
<section id="ancestral-state-estimation-using-phytools" class="level2">
<h2 class="anchored" data-anchor-id="ancestral-state-estimation-using-phytools">7.0 Ancestral state estimation using Phytools</h2>
<p>Phytools supports a wide variety of ancestral state estimation. There are plenty to choose from and I’d do some reading around to determine which models will work best for you. I’ve stored a couple of tutorials in the session folder for the estimation of continuous ancestral states under Brownian motion and discrete ancestral states under the Mk model as HTML documents. It might be worth looking through these in your spare time. In the meantime, we’ll go through ancr, a function which lets you estimate ancestral states under almost any model, and Bayesian stochastic character mapping.</p>
<p>As stated previously, you can should estimate your ancestral states prior to ordination and distance matrix derivation (if applicable), add these values to your main data matrix, and then just analyse the supplemented matrix as a normal matrix. Note that geomorph let’s you do this with GPA-aligned coordinates. This is absolutely fine.</p>
<section id="ancr" class="level3">
<h3 class="anchored" data-anchor-id="ancr">7.1 ancr</h3>
<p>The idea of this (generic) method is that it takes a fitted discrete character evolution (e.g., from fitMk or fitHRM) and then computes the marginal reconstructions under that model. Marginal ancestral states are also referred to as scaled likelihoods, and it’s valid to interpret them as the probabilities that each node is in each of the different states, conditioning on the fitted model. (This is why they are also called “empirical Bayes” posterior probabilities – because they are posterior probabilities but in which we condition, normally, on the ML value of our transition model, Q.)</p>
<p>To illustrate this method, we’re going to apply it to a trait evolution scenario for the so-called “hidden-rates-model” of Beaulieu et al.&nbsp;(2013). For this, we’ll use a dataset of parity mode in liolaemid lizards from Esquerré et al.&nbsp;(2018). Previous analyses found that the best-supported model for these data was a class of hidden-rates-model that I refer to as the ‘umbral’ model in which we have hidden states of the character (say 0* and 1* for an observed character with two levels 0 &amp; 1), but in which evolution always had to proceed 0* ⇆ 0 ⇆ 1 ⇆ 1*. The idea is to capture “cold” and “hot” conditions for the state, in which when in the cold condition the character can’t evolve at all without first transitioning back to the hot condition.</p>
<p>Load the tree.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>liolaemid.tree<span class="ot">&lt;-</span><span class="fu">read.nexus</span>(</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">file=</span><span class="st">"http://www.phytools.org/Rbook/7/Liolaemidae.MCC.nex"</span>)</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>liolaemid.tree<span class="ot">&lt;-</span><span class="fu">untangle</span>(<span class="fu">ladderize</span>(liolaemid.tree),<span class="st">"read.tree"</span>)</span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>liolaemid.tree</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 258 tips and 257 internal nodes.

Tip labels:
  Liolaemus_albiceps, Liolaemus_irregularis, Liolaemus_ornatus, Liolaemus_calchaqui, Liolaemus_crepuscularis, Liolaemus_lavillai, ...

Rooted; includes branch lengths.</code></pre>
</div>
</div>
<p>Load the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>liolaemid.data<span class="ot">&lt;-</span><span class="fu">read.csv</span>(</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">file=</span><span class="st">"http://www.phytools.org/Rbook/7/Liolaemidae.data.csv"</span>,</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names=</span><span class="dv">1</span>,<span class="at">stringsAsFactors=</span><span class="cn">TRUE</span>)</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>parity_mode<span class="ot">&lt;-</span><span class="fu">setNames</span>(liolaemid.data<span class="sc">$</span>parity_mode,</span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(liolaemid.data))</span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(parity_mode)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"oviparous"</span>,<span class="st">"viviparous"</span>)</span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(parity_mode)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Ctenoblepharys_adspersa       Liolaemus_abaucan      Liolaemus_albiceps 
              oviparous               oviparous              viviparous 
      Liolaemus_andinus     Liolaemus_annectens      Liolaemus_anomalus 
             viviparous              viviparous               oviparous 
Levels: oviparous viviparous</code></pre>
</div>
</div>
<p>So far so good. Now, we’ll fit our model using phytools::fitHRM. Be forewarned: this takes a while!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>hrm.umbral<span class="ot">&lt;-</span><span class="fu">fitHRM</span>(liolaemid.tree,parity_mode,<span class="at">umbral=</span><span class="cn">TRUE</span>,</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">pi=</span><span class="st">"fitzjohn"</span>,<span class="at">opt.method=</span><span class="st">"optimParallel"</span>,<span class="at">rand_start=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
This is the design matrix of the fitted model.
Does it make sense?

            oviparous oviparous* viviparous viviparous*
oviparous           0          1          2           0
oviparous*          3          0          0           0
viviparous          4          0          0           5
viviparous*         0          0          6           0</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in log(comp[1:M + N]): NaNs produced</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>log-likelihood from current iteration: -63.5918 
 --- Best log-likelihood so far: -63.5918 ---</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in log(comp[1:M + N]): NaNs produced</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>log-likelihood from current iteration: -65.9156 
 --- Best log-likelihood so far: -63.5918 ---</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in log(comp[1:M + N]): NaNs produced</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>log-likelihood from current iteration: -63.4594 
 --- Best log-likelihood so far: -63.4594 ---
log-likelihood from current iteration: -65.9463 
 --- Best log-likelihood so far: -63.4594 ---</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in log(comp[1:M + N]): NaNs produced</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>log-likelihood from current iteration: -63.592 
 --- Best log-likelihood so far: -63.4594 ---
log-likelihood from current iteration: -60.6925 
 --- Best log-likelihood so far: -60.6925 ---
log-likelihood from current iteration: -65.916 
 --- Best log-likelihood so far: -60.6925 ---</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in log(comp[1:M + N]): NaNs produced</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>log-likelihood from current iteration: -65.947 
 --- Best log-likelihood so far: -60.6925 ---</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in log(comp[1:M + N]): NaNs produced</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>log-likelihood from current iteration: -63.4594 
 --- Best log-likelihood so far: -60.6925 ---</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in log(comp[1:M + N]): NaNs produced</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>log-likelihood from current iteration: -65.9156 
 --- Best log-likelihood so far: -60.6925 ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>hrm.umbral</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Object of class "fitHRM".

Observed states: [ oviparous, viviparous ]
Number of rate categories per state: [ 2, 2 ]

Fitted (or set) value of Q:
            oviparous oviparous* viviparous viviparous*
oviparous   -1.962662   0.039898   1.922764     0.00000
oviparous*   0.037901  -0.037901   0.000000     0.00000
viviparous   2.684883   0.000000  -3.099423     0.41454
viviparous*  0.000000   0.000000   0.000000     0.00000

Fitted (or set) value of pi:
  oviparous  oviparous*  viviparous viviparous* 
   0.034102    0.939368    0.026530    0.000000 
due to treating the root prior as (a) nuisance.

Log-likelihood: -60.692453 

Optimization method used was "optimParallel"

R thinks it has found the ML solution.</code></pre>
</div>
</div>
<p>If we graph our fitted model, we should see that it corresponds nicely to the process we hypothesized.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hrm.umbral,<span class="at">spacer=</span><span class="fl">0.3</span>,<span class="at">offset=</span><span class="fl">0.03</span>,</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">mar=</span><span class="fu">rep</span>(<span class="fl">0.1</span>,<span class="dv">4</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Having fit this model to our data, we just need to pass the fitted object directly to our new ancr (formerly, briefly anceb) as follows. The only argument I’m going to specify is tips=TRUE which will return the posterior probabilities that each of the leaves of the tree are in each of the two unobserved conditions of the character, given its observed state.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>parity_ancr<span class="ot">&lt;-</span><span class="fu">ancr</span>(hrm.umbral,<span class="at">tips=</span><span class="cn">TRUE</span>)</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>parity_ancr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Marginal ancestral state estimates:
    oviparous oviparous* viviparous viviparous*
259  0.001315   0.997889   0.000796           0
260  0.020644   0.968556   0.010800           0
261  0.023006   0.965186   0.011808           0
262  0.015974   0.976061   0.007966           0
263  0.085224   0.867177   0.047599           0
264  0.071613   0.893705   0.034682           0
...

Log-likelihood = -60.692453 </code></pre>
</div>
</div>
<p>Lastly, let’s graph these probabilities onto the tree. This is a pretty swish bit of plotting code courtesy of Liam Revell’s blog (a place I’d highly recommend visiting).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set max node height</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>h<span class="ot">&lt;-</span><span class="fu">max</span>(<span class="fu">nodeHeights</span>(liolaemid.tree))</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot</span></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTree</span>(liolaemid.tree,<span class="at">ftype=</span><span class="st">"off"</span>,<span class="at">lwd=</span><span class="dv">1</span>,</span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">1.05</span><span class="sc">*</span>h),<span class="at">direction=</span><span class="st">"upwards"</span>)</span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>pp<span class="ot">&lt;-</span><span class="fu">get</span>(<span class="st">"last_plot.phylo"</span>,<span class="at">envir=</span>.PlotPhyloEnv)</span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>cols<span class="ot">&lt;-</span><span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">"#f0ead6"</span>,<span class="st">"#fffff2"</span>,<span class="st">"darkred"</span>,<span class="fu">palette</span>()[<span class="dv">2</span>]),</span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(parity_ancr<span class="sc">$</span>ace))</span>
<span id="cb103-9"><a href="#cb103-9" aria-hidden="true" tabindex="-1"></a><span class="co"># plot boxes along top of tree</span></span>
<span id="cb103-10"><a href="#cb103-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">Ntip</span>(liolaemid.tree)){</span>
<span id="cb103-11"><a href="#cb103-11" aria-hidden="true" tabindex="-1"></a>  tcp<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fu">cumsum</span>(parity_ancr<span class="sc">$</span>ace[liolaemid.tree<span class="sc">$</span>tip.label[i],]))</span>
<span id="cb103-12"><a href="#cb103-12" aria-hidden="true" tabindex="-1"></a>  tcp<span class="ot">&lt;-</span>tcp<span class="sc">*</span><span class="fl">0.05</span><span class="sc">*</span>h</span>
<span id="cb103-13"><a href="#cb103-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(tcp)){</span>
<span id="cb103-14"><a href="#cb103-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">polygon</span>(<span class="fu">rep</span>(pp<span class="sc">$</span>xx[i],<span class="dv">4</span>)<span class="sc">+</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.5</span>,<span class="sc">-</span><span class="fl">0.5</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>),</span>
<span id="cb103-15"><a href="#cb103-15" aria-hidden="true" tabindex="-1"></a>      h<span class="sc">+</span><span class="fu">c</span>(tcp[j<span class="dv">-1</span>],tcp[j],tcp[j],tcp[j<span class="dv">-1</span>]),</span>
<span id="cb103-16"><a href="#cb103-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">border=</span><span class="cn">FALSE</span>,<span class="at">col=</span>cols[j<span class="dv">-1</span>])</span>
<span id="cb103-17"><a href="#cb103-17" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb103-18"><a href="#cb103-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb103-19"><a href="#cb103-19" aria-hidden="true" tabindex="-1"></a><span class="co"># set legend</span></span>
<span id="cb103-20"><a href="#cb103-20" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="at">x=</span><span class="dv">0</span>,<span class="at">y=</span><span class="fl">0.5</span><span class="sc">*</span>h,<span class="fu">colnames</span>(parity_ancr<span class="sc">$</span>ace),<span class="at">pch=</span><span class="dv">15</span>,</span>
<span id="cb103-21"><a href="#cb103-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">col=</span>cols,<span class="at">pt.cex=</span><span class="fl">1.5</span>,<span class="at">cex=</span><span class="fl">0.8</span>,<span class="at">bty=</span><span class="st">"n"</span>)</span>
<span id="cb103-22"><a href="#cb103-22" aria-hidden="true" tabindex="-1"></a><span class="co"># set node labels</span></span>
<span id="cb103-23"><a href="#cb103-23" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">fg=</span><span class="st">"transparent"</span>)</span>
<span id="cb103-24"><a href="#cb103-24" aria-hidden="true" tabindex="-1"></a><span class="fu">nodelabels</span>(<span class="at">pie=</span>parity_ancr<span class="sc">$</span>ace[<span class="dv">1</span><span class="sc">:</span><span class="fu">Nnode</span>(liolaemid.tree)<span class="sc">+</span></span>
<span id="cb103-25"><a href="#cb103-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Ntip</span>(liolaemid.tree),],<span class="at">piecol=</span>cols,<span class="at">cex=</span><span class="fl">0.3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="stochastic-character-mapping" class="level3">
<h3 class="anchored" data-anchor-id="stochastic-character-mapping">7.2 stochastic character mapping</h3>
<p>Stochastic character mapping is a fully Bayesian approach to ancestral state estimation. Phytools supports this via the make.simmap function. This is my preferred option for the ancestral state estimation of discrete characters.</p>
<p>Generate tree and Q matrix (defining substitution rates), then add character data to tree.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">pbtree</span>(<span class="at">n =</span> <span class="dv">200</span>, <span class="at">scale =</span> <span class="dv">1</span>)</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>Q <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="sc">-</span><span class="dv">1</span>),<span class="dv">2</span>,<span class="dv">2</span>)</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(Q) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(Q) <span class="ot">&lt;-</span> letters[<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(Q)]</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">sim.history</span>(tree,Q,<span class="at">anc=</span><span class="st">"a"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Done simulation(s).</code></pre>
</div>
</div>
<p>Now conduct stochastic character mapping.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>mtrees <span class="ot">&lt;-</span> <span class="fu">make.simmap</span>(tree, tree<span class="sc">$</span>states, <span class="at">model=</span><span class="st">"ER"</span>, <span class="at">nsim =</span> <span class="dv">100</span>, <span class="at">pi=</span><span class="st">"estimated"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Using pi estimated from the stationary distribution of Q assuming a flat prior.
pi =
  a   b 
0.5 0.5 

make.simmap is sampling character histories conditioned on
the transition matrix

Q =
          a         b
a -1.034528  1.034528
b  1.034528 -1.034528
(estimated using likelihood);
and (mean) root node prior probabilities
pi =
  a   b 
0.5 0.5 </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done.</code></pre>
</div>
</div>
<p>Now let’s plot the first result using the bespoke function phytools provides. These results are binary because it’s just one tree.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">"blue"</span>,<span class="st">"red"</span>), letters[<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(Q)])</span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSimmap</span>(mtrees[[<span class="dv">1</span>]], cols , <span class="at">pts=</span>F , <span class="at">ftype=</span><span class="st">"off"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>One advantage of stochastic character mapping is that we can compute the posterior probabilities of each character state being present at each node. Let’s summarise across all the simmaps we produced. First we’ll define a function that will do this for us.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co"># function to compute the states</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a> foo <span class="ot">&lt;-</span> <span class="cf">function</span>(x){</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a> y <span class="ot">&lt;-</span> <span class="fu">sapply</span>(x<span class="sc">$</span>maps,<span class="cf">function</span>(x) <span class="fu">names</span>(x)[<span class="dv">1</span>])</span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a> <span class="fu">names</span>(y) <span class="ot">&lt;-</span> x<span class="sc">$</span>edge[,<span class="dv">1</span>]</span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a> y <span class="ot">&lt;-</span> y[<span class="fu">as.character</span>(<span class="fu">length</span>(x<span class="sc">$</span>tip)<span class="sc">+</span><span class="dv">1</span><span class="sc">:</span>x<span class="sc">$</span>Nnode)]</span>
<span id="cb110-6"><a href="#cb110-6" aria-hidden="true" tabindex="-1"></a> <span class="fu">return</span>(y)</span>
<span id="cb110-7"><a href="#cb110-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb110-8"><a href="#cb110-8" aria-hidden="true" tabindex="-1"></a><span class="co"># now apply the function</span></span>
<span id="cb110-9"><a href="#cb110-9" aria-hidden="true" tabindex="-1"></a>AA <span class="ot">&lt;-</span> <span class="fu">sapply</span>(mtrees,foo)</span>
<span id="cb110-10"><a href="#cb110-10" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to data that be plotted as pie charts.</span></span>
<span id="cb110-11"><a href="#cb110-11" aria-hidden="true" tabindex="-1"></a>piesA <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(AA,<span class="dv">1</span>,<span class="cf">function</span>(x,levels,Nsim) <span class="fu">summary</span>(<span class="fu">factor</span>(x,levels))<span class="sc">/</span>Nsim,<span class="at">levels=</span>letters[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">Nsim=</span><span class="dv">100</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s plot the summary of our analyses. This is a fun way of see how characters evolve along a tree.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.phylo</span>(tree,<span class="at">no.margin=</span><span class="cn">TRUE</span>,<span class="at">show.tip.label=</span>F)</span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="fu">nodelabels</span>(<span class="at">pie=</span>piesA,<span class="at">cex=</span><span class="fl">0.6</span>,<span class="at">piecol=</span>cols)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Ordination_and_phylogenetic_comparative_methods_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In your disparity analyses, you will have to decide on a probability threshold for determining character scores at each node. This is entirely arbitrary and can be worth exploring.</p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>